{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ee350533",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "\n",
    "import evo_helpers as hlp\n",
    "import evo_constants as CTS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c35ea934",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome to EvoMinig\n",
      "\n"
     ]
    }
   ],
   "source": [
    "hlp.welcome()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c37a782c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Press 1 to continue with precharged example.\n",
      "Press 2 to continue with your own databases.\n"
     ]
    }
   ],
   "source": [
    "#from importlib import reload\n",
    "#reload(evo_helpers)\n",
    "\n",
    "opt = hlp.option_menu()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "fd17f884",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Evomining:\n",
    "    \"\"\"This class creates an evomining object.\n",
    "    Its methods permit to access to the evomining\n",
    "    funtionality.\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, centralDB,evo_genomesDB, nat_prodsDB):\n",
    "        self.central = centralDB\n",
    "        self.genomesDB = evo_genomesDB\n",
    "        self.nat_prods = nat_prodsDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6fc4b6a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "## obtener las bases de datos del usuario\n",
    "\n",
    "#from importlib import reload\n",
    "#reload(evo_helpers)\n",
    "\n",
    "## option 1 precharged example:\n",
    "if opt == 1:\n",
    "    hlp.make_all_evo_headers(CTS.EXMPL_DB[\"genomes\"],\n",
    "                                 CTS.EXMPL_DB[\"rast_ids\"])\n",
    "    hlp.join_evo_headers_files()\n",
    "    hlp.remove_evo_headers_files()\n",
    "    \n",
    "    evo_genomes_path = CTS.EVO_GENOMES_DB\n",
    "    \n",
    "    #evomining = EvoMining(cts.EXML_DB[\"central\"],\n",
    "    #                      cts.EXMPL_DB[\"genomes\"],\n",
    "    #                      cts.EXMPL_DB[\"nat_prods\"])\n",
    "    #evomining.run()\n",
    "## option 2 user databases:\n",
    "else: \n",
    "    user_dbs = hlp.get_dbs()\n",
    "    hlp.make_all_evo_headers(user_dbs[\"genomes\"], \n",
    "                             user_dbs[\"rast_ids\"])\n",
    "    hlp.join_evo_headers_files()\n",
    "    hlp.remove_evo_headers_files()\n",
    "    \n",
    "    user_dbs[\"genomes\"] = CTS.EVO_GENOMES_DB\n",
    "    \n",
    "    evomining = EvoMining(user_dbs[\"central\"],\n",
    "                         user_dbs[\"genomes\"],\n",
    "                         user_dbs[\"nat_prods\"])\n",
    "    #evomining.run()\n",
    "\n",
    "#evomining.run()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b349be4",
   "metadata": {},
   "source": [
    "## Obtener los archivos .faa con los encabezados en formato evomining:\n",
    "### Los nuevos .faa individuales se guardan en data/aux_data/  (se borran una vez utilizados)\n",
    "### La base de datos con encabezados de evomining se guarda en /data/data_bases/evo_genomes_db/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "696da2e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#from importlib import reload\n",
    "#reload(evo_helpers)\n",
    "\n",
    "genomes_db = \"/home/luisyovanny/.evo/dev_Evo/data/data_bases/example/genomes/los17\"\n",
    "Rast_ids = \"/home/luisyovanny/.evo/dev_Evo/data/data_bases/example/genomes/los17Rast.ids\"\n",
    "hlp.make_all_evo_headers(genomes_db, Rast_ids)\n",
    "hlp.join_evo_headers_files()\n",
    "#hlp.remove_evo_headers_files()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8d0f4d3",
   "metadata": {},
   "source": [
    "## Aplicar makeblastdb a las bases de datos que necesitaremos:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "996103de",
   "metadata": {},
   "outputs": [],
   "source": [
    "hlp.makeblast_db(CTS.EVO_GENOMES_DB)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "942d751a",
   "metadata": {},
   "outputs": [],
   "source": [
    "hlp.makeblast_db(CTS.EXMPL_DB[\"central\"])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "292f2df2",
   "metadata": {},
   "outputs": [],
   "source": [
    "hlp.makeblast_db(CTS.EXMPL_DB[\"nat_prods\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53e73461",
   "metadata": {},
   "source": [
    "## Obtener los blastp entre la base de datos central vs la base de datos de genomas:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "da5384c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "blastdb_path = CTS.BLASTDBs_PATH + 'evo_genomes_db_blastdb/'\n",
    "            \n",
    "hlp.apply_blastp(CTS.EXMPL_DB[\"central\"], blastdb_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ebc9aa70",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/luisyovanny/.evo/dev_Evo/data/data_bases/evo_genomes_db/'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CTS.EVO_GENOMES_DB"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e22b24f",
   "metadata": {},
   "source": [
    "## Obtener los blastp entre la base de datos de genomas vs la base de datos central:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "70239247",
   "metadata": {},
   "outputs": [],
   "source": [
    "blastdb_path = CTS.BLASTDBs_PATH + 'central_blastdb/'\n",
    "            \n",
    "hlp.apply_blastp(CTS.EVO_GENOMES_DB, blastdb_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1b1b9ac",
   "metadata": {},
   "source": [
    "## Obtener los BBH:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e78c3eae",
   "metadata": {},
   "source": [
    "### Obtener los mejores elementos únicos del blast de metabolismo central vs genomas: \n",
    "### Estos se guardarán en un diccionario de la forma:\n",
    "### { enzima_i : mejor_copia_en_el_organismo_k }\n",
    "### donde i recorre todas las enzimas en el metabolismo central, y k recorre todos los genomas.\n",
    "### - Considerar solo con bitscore >= 100 e E_value < 0.001 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "791b41e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "df1 = pd.read_csv('/home/luisyovanny/.evo/dev_Evo/data/data_bases/blastp/central_to_genomes/central_to_genomes.blast', sep = \"\\t\", names = CTS.BLAST_COLS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "d65a289e",
   "metadata": {},
   "outputs": [],
   "source": [
    "central_enzimes = df1['query'].unique() ##Lista con las enzimas del metabolismo central\n",
    "genomes = hlp.dict_ids_to_names_genomes(CTS.EXMPL_DB[\"rast_ids\"]) ## Diccio con los genomas de la base de genomas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "6fcfd9fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_1|Cglu',\n",
       "       '3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_2|Mtub',\n",
       "       '3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_3|Mtub',\n",
       "       '3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_4|Scoe',\n",
       "       '3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_1|none',\n",
       "       '3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_2|none',\n",
       "       '3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_3|none'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "central_enzimes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "8fd9e030",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['1156913.7', '169133.3', '290340.24', '367928.21', '442563.11', '446462.15', '452863.24', '649831.6', '6666666.104540', '6666666.111158', '6666666.111572', '6666666.112811', '6666666.146813', '6666666.146852', '6666666.146872', '749927.13', '930171.11'])"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genomes.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "d5bc09ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "#definimos el diccionario que será la salida:\n",
    "unique_in_first_blast = dict()\n",
    "#inicializamos el diccionario tomando por claves todas las enzimas\n",
    "# y por el momento asignando por valores listas vacías:\n",
    "for enzime in central_enzimes:\n",
    "    unique_in_first_blast[enzime] = []\n",
    "    \n",
    "#definimos una lista auxiliar que nos permitirá quedarnos con la mejor\n",
    "# copia de la enzima presente en cada organismo \n",
    "\n",
    "#aux_list = list()\n",
    "for enzime in central_enzimes:\n",
    "    for genome_id in genomes.keys():\n",
    "        aux_list = []\n",
    "        if genome_id != \"6666666.385189\":\n",
    "            for index in range(len(df1)):\n",
    "                if (aux_list == []):\n",
    "                    if ((enzime == df1[\"query\"][index])&(genome_id in df1[\"subject\"][index])):\n",
    "                        aux_list = list(df1.iloc[index])\n",
    "                    else:\n",
    "                        continue\n",
    "                else:\n",
    "                    if ((enzime == df1[\"query\"][index]) & (genome_id in df1[\"subject\"][index]) & (df1[\"bitscore\"][index] > aux_list[11])):\n",
    "                        aux_list = list(df1.iloc[index])\n",
    "            if aux_list != []:  \n",
    "                unique_in_first_blast[enzime].append(aux_list)\n",
    "                \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "64f9f18e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_3|none',\n",
       "  'gi|1156913.7.7342|1156913.7|NC_021252.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Amycolatopsis_orientalis_HCCB10007_NC_021252.1',\n",
       "  67.769,\n",
       "  363,\n",
       "  116,\n",
       "  1,\n",
       "  1,\n",
       "  363,\n",
       "  13,\n",
       "  374,\n",
       "  3.0300000000000004e-179,\n",
       "  503.0],\n",
       " ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_3|none',\n",
       "  'gi|169133.3.2262|169133.3|LAB|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Citricoccus_sp_CH26_LAB',\n",
       "  68.32,\n",
       "  363,\n",
       "  115,\n",
       "  0,\n",
       "  1,\n",
       "  363,\n",
       "  12,\n",
       "  374,\n",
       "  0.0,\n",
       "  524.0],\n",
       " ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_3|none',\n",
       "  'gi|290340.24.1033|290340.24|NC_008711.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Arthrobacter_aurescens_TC1_NC_008711.1',\n",
       "  67.493,\n",
       "  363,\n",
       "  118,\n",
       "  0,\n",
       "  1,\n",
       "  363,\n",
       "  12,\n",
       "  374,\n",
       "  0.0,\n",
       "  519.0],\n",
       " ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_3|none',\n",
       "  'gi|367928.21.407|367928.21|NC_008618.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Bifidobacterium_adolescentis_ATCC_15703_NC_008618.1',\n",
       "  57.297,\n",
       "  370,\n",
       "  151,\n",
       "  2,\n",
       "  1,\n",
       "  363,\n",
       "  11,\n",
       "  380,\n",
       "  8.969999999999999e-152,\n",
       "  433.0],\n",
       " ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_3|none',\n",
       "  'gi|442563.11.458|442563.11|AD011|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Bifidobacterium_animalis_subsp._lactis_AD011',\n",
       "  53.243,\n",
       "  370,\n",
       "  166,\n",
       "  2,\n",
       "  1,\n",
       "  363,\n",
       "  10,\n",
       "  379,\n",
       "  1.53e-142,\n",
       "  409.0],\n",
       " ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_3|none',\n",
       "  'gi|446462.15.520|446462.15|NC_021252.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Actinosynnema_mirum_DSM_43827_NC_013093.1_NC_021252.1',\n",
       "  66.942,\n",
       "  363,\n",
       "  119,\n",
       "  1,\n",
       "  1,\n",
       "  363,\n",
       "  16,\n",
       "  377,\n",
       "  0.0,\n",
       "  514.0],\n",
       " ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_3|none',\n",
       "  'gi|452863.24.923|452863.24|NC_011886.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Arthrobacter_chlorophenolicus_A6_NC_011886.1',\n",
       "  68.32,\n",
       "  363,\n",
       "  115,\n",
       "  0,\n",
       "  1,\n",
       "  363,\n",
       "  12,\n",
       "  374,\n",
       "  0.0,\n",
       "  524.0],\n",
       " ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_3|none',\n",
       "  'gi|649831.6.513|649831.6|NC_021191.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Actinoplanes_sp._N902-109_NC_021191.1',\n",
       "  62.81,\n",
       "  363,\n",
       "  135,\n",
       "  0,\n",
       "  1,\n",
       "  363,\n",
       "  11,\n",
       "  373,\n",
       "  4.499999999999999e-170,\n",
       "  479.0],\n",
       " ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_3|none',\n",
       "  'gi|6666666.104540.2193|6666666.104540|AOHO01|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Amycolatopsis_decaplanina_DSM_44594_AOHO01',\n",
       "  67.769,\n",
       "  363,\n",
       "  116,\n",
       "  1,\n",
       "  1,\n",
       "  363,\n",
       "  13,\n",
       "  374,\n",
       "  0.0,\n",
       "  521.0],\n",
       " ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_3|none',\n",
       "  'gi|6666666.111158.8940|6666666.111158|NC_0131311|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Catenulispora_acidiphila_DSM_44928_NC_0131311',\n",
       "  73.003,\n",
       "  363,\n",
       "  98,\n",
       "  0,\n",
       "  1,\n",
       "  363,\n",
       "  10,\n",
       "  372,\n",
       "  0.0,\n",
       "  564.0],\n",
       " ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_3|none',\n",
       "  'gi|6666666.111572.2478|6666666.111572|NC_0094801|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Clavibacter_michiganensis_subsp._michiganensis_NCPPB_382_NC_0094801',\n",
       "  64.641,\n",
       "  362,\n",
       "  126,\n",
       "  2,\n",
       "  1,\n",
       "  362,\n",
       "  302,\n",
       "  661,\n",
       "  1.79e-158,\n",
       "  460.0],\n",
       " ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_3|none',\n",
       "  'gi|6666666.112811.7991|6666666.112811|JOEF01|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Allokutzneria_albata_JOEF01',\n",
       "  67.218,\n",
       "  363,\n",
       "  118,\n",
       "  1,\n",
       "  1,\n",
       "  363,\n",
       "  16,\n",
       "  377,\n",
       "  0.0,\n",
       "  513.0],\n",
       " ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_3|none',\n",
       "  'gi|6666666.146813.572|6666666.146813|NC_014550.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Arthrobacter_arilaitensis_RE117_NC_014550.1',\n",
       "  61.708,\n",
       "  363,\n",
       "  139,\n",
       "  0,\n",
       "  1,\n",
       "  363,\n",
       "  11,\n",
       "  373,\n",
       "  6.15e-169,\n",
       "  476.0],\n",
       " ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_3|none',\n",
       "  'gi|6666666.146852.774|6666666.146852|NC_002935.2|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Corynebacterium_diphtheriae_NCTC_13129_NC_002935.2',\n",
       "  61.326,\n",
       "  362,\n",
       "  139,\n",
       "  1,\n",
       "  1,\n",
       "  362,\n",
       "  11,\n",
       "  371,\n",
       "  7.62e-167,\n",
       "  471.0],\n",
       " ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_3|none',\n",
       "  'gi|6666666.146872.438|6666666.146872|NC_017803.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Actinoplanes_sp._SE50/110_NC_017803.1',\n",
       "  60.055,\n",
       "  363,\n",
       "  145,\n",
       "  0,\n",
       "  1,\n",
       "  363,\n",
       "  12,\n",
       "  374,\n",
       "  1.28e-166,\n",
       "  471.0],\n",
       " ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_3|none',\n",
       "  'gi|749927.13.8695|749927.13|NC_014318.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Amycolatopsis_mediterranei_U32_NC_014318.1',\n",
       "  68.871,\n",
       "  363,\n",
       "  112,\n",
       "  1,\n",
       "  1,\n",
       "  363,\n",
       "  12,\n",
       "  373,\n",
       "  0.0,\n",
       "  527.0],\n",
       " ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_3|none',\n",
       "  'gi|930171.11.818|930171.11|NC_015145.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Arthrobacter_phenanthrenivorans_Sphe3_NC_015145.1',\n",
       "  67.218,\n",
       "  363,\n",
       "  119,\n",
       "  0,\n",
       "  1,\n",
       "  363,\n",
       "  12,\n",
       "  374,\n",
       "  0.0,\n",
       "  518.0]]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_in_first_blast[\"3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_3|none\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "abdf3a65",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "119"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total = 0\n",
    "for enz in central_enzimes:\n",
    "        total = total + len(unique_in_first_blast[enz])\n",
    "total"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a054be1",
   "metadata": {},
   "source": [
    "### NOTA: En la base de datos de genomas está de más el genoma con id = 6666666.385189"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f060dbd",
   "metadata": {},
   "source": [
    "Unique de Genomes vs EF:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "f45f9e38",
   "metadata": {},
   "outputs": [],
   "source": [
    "df2 = pd.read_csv('/home/luisyovanny/.evo/dev_Evo/data/data_bases/blastp/genomes_to_central/genomes_to_central.blast', sep = \"\\t\", names = CTS.BLAST_COLS )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "17519f45",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>subject</th>\n",
       "      <th>pc_identity</th>\n",
       "      <th>aln_length</th>\n",
       "      <th>mismatches</th>\n",
       "      <th>gaps_opened</th>\n",
       "      <th>query_start</th>\n",
       "      <th>query_end</th>\n",
       "      <th>subject_start</th>\n",
       "      <th>subject_end</th>\n",
       "      <th>e_value</th>\n",
       "      <th>bitscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gi|169133.3.1663|169133.3|LAB|Homoserine_dehyd...</td>\n",
       "      <td>3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogena...</td>\n",
       "      <td>37.778</td>\n",
       "      <td>45</td>\n",
       "      <td>28</td>\n",
       "      <td>0</td>\n",
       "      <td>342</td>\n",
       "      <td>386</td>\n",
       "      <td>442</td>\n",
       "      <td>486</td>\n",
       "      <td>3.230000e-06</td>\n",
       "      <td>37.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gi|169133.3.2262|169133.3|LAB|Phosphoserine_am...</td>\n",
       "      <td>3PGA_AMINOACIDS|2|Phosphoserine_aminotransfera...</td>\n",
       "      <td>68.320</td>\n",
       "      <td>363</td>\n",
       "      <td>115</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>374</td>\n",
       "      <td>1</td>\n",
       "      <td>363</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>524.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gi|169133.3.2262|169133.3|LAB|Phosphoserine_am...</td>\n",
       "      <td>3PGA_AMINOACIDS|2|Phosphoserine_aminotransfera...</td>\n",
       "      <td>64.785</td>\n",
       "      <td>372</td>\n",
       "      <td>130</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>374</td>\n",
       "      <td>6</td>\n",
       "      <td>376</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>504.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gi|169133.3.2262|169133.3|LAB|Phosphoserine_am...</td>\n",
       "      <td>3PGA_AMINOACIDS|2|Phosphoserine_aminotransfera...</td>\n",
       "      <td>61.202</td>\n",
       "      <td>366</td>\n",
       "      <td>141</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>372</td>\n",
       "      <td>6</td>\n",
       "      <td>370</td>\n",
       "      <td>3.710000e-171</td>\n",
       "      <td>473.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gi|169133.3.2437|169133.3|LAB|putative_2-hydro...</td>\n",
       "      <td>3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogena...</td>\n",
       "      <td>53.261</td>\n",
       "      <td>92</td>\n",
       "      <td>43</td>\n",
       "      <td>0</td>\n",
       "      <td>211</td>\n",
       "      <td>302</td>\n",
       "      <td>195</td>\n",
       "      <td>286</td>\n",
       "      <td>3.370000e-21</td>\n",
       "      <td>82.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               query  \\\n",
       "0  gi|169133.3.1663|169133.3|LAB|Homoserine_dehyd...   \n",
       "1  gi|169133.3.2262|169133.3|LAB|Phosphoserine_am...   \n",
       "2  gi|169133.3.2262|169133.3|LAB|Phosphoserine_am...   \n",
       "3  gi|169133.3.2262|169133.3|LAB|Phosphoserine_am...   \n",
       "4  gi|169133.3.2437|169133.3|LAB|putative_2-hydro...   \n",
       "\n",
       "                                             subject  pc_identity  aln_length  \\\n",
       "0  3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogena...       37.778          45   \n",
       "1  3PGA_AMINOACIDS|2|Phosphoserine_aminotransfera...       68.320         363   \n",
       "2  3PGA_AMINOACIDS|2|Phosphoserine_aminotransfera...       64.785         372   \n",
       "3  3PGA_AMINOACIDS|2|Phosphoserine_aminotransfera...       61.202         366   \n",
       "4  3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogena...       53.261          92   \n",
       "\n",
       "   mismatches  gaps_opened  query_start  query_end  subject_start  \\\n",
       "0          28            0          342        386            442   \n",
       "1         115            0           12        374              1   \n",
       "2         130            1            3        374              6   \n",
       "3         141            1            7        372              6   \n",
       "4          43            0          211        302            195   \n",
       "\n",
       "   subject_end        e_value  bitscore  \n",
       "0          486   3.230000e-06      37.7  \n",
       "1          363   0.000000e+00     524.0  \n",
       "2          376   0.000000e+00     504.0  \n",
       "3          370  3.710000e-171     473.0  \n",
       "4          286   3.370000e-21      82.0  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "246b9cc4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "17"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(genomes.keys())\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "768403a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#definimos el diccionario que será la salida:\n",
    "unique_in_second_blast = dict()\n",
    "#inicializamos el diccionario tomando por claves todas las enzimas\n",
    "# y por el momento asignando por valores listas vacías:\n",
    "for genome_id in genomes.keys():\n",
    "    if genome_id != \"6666666.385189\":\n",
    "        unique_in_second_blast[genome_id] = []\n",
    "    else:\n",
    "        continue\n",
    "\n",
    "#definimos una lista auxiliar que nos permitirá quedarnos con la mejor\n",
    "# copia de la enzima presente en cada organismo \n",
    "\n",
    "\n",
    "for genome_id in genomes.keys():\n",
    "    if genome_id != \"6666666.385189\":\n",
    "        for enzime in central_enzimes:\n",
    "            aux_list = []\n",
    "            for index in range(len(df2)):\n",
    "                if (aux_list == []):\n",
    "                    if ((genome_id in df2[\"query\"][index])&(enzime in df2[\"subject\"][index])):\n",
    "                        aux_list = list(df2.iloc[index])\n",
    "                    else:\n",
    "                        continue\n",
    "                else:\n",
    "                    if ((genome_id in df2[\"query\"][index]) & (enzime in df2[\"subject\"][index]) & (df2[\"bitscore\"][index] > aux_list[11])):\n",
    "                    \n",
    "                        aux_list = list(df2.iloc[index])\n",
    "            if aux_list != []:  \n",
    "                unique_in_second_blast[genome_id].append(aux_list)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e2cfb07",
   "metadata": {},
   "source": [
    "### Prueba de bio.genbank package:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "672cc1c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "from taxonomy_ranks import TaxonomyRanks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "9ebbadf1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "NCBI database not present yet (first time used?)\n",
      "Downloading taxdump.tar.gz from NCBI FTP site (via HTTP)...\n",
      "Done. Parsing...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading node names...\n",
      "2609043 names loaded.\n",
      "340117 synonyms loaded.\n",
      "Loading nodes...\n",
      "2609043 nodes loaded.\n",
      "Linking nodes...\n",
      "Tree is loaded.\n",
      "Updating database: /home/luisyovanny/.etetoolkit/taxa.sqlite ...\n",
      " 2609000 generating entries... "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Inserting synonyms:      20000 "
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Uploading to /home/luisyovanny/.etetoolkit/taxa.sqlite\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Inserting synonyms:      340000 "
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Inserting taxids:       45000  "
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Inserting taxids:       2605000  "
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "taxa_name = 'Actinomadura hibisca'\n",
    "\n",
    "rank_taxon = TaxonomyRanks(taxa_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "58404e86",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "taxonomy_ranks.get_taxonomy_rank_with_ete3_with_super_and_sub.TaxonomyRanks"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(rank_taxon)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "58c4f3ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "rank_taxon.get_lineage_taxids_and_taxanames()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "67fd8f50",
   "metadata": {},
   "outputs": [],
   "source": [
    "ranks = ('user_taxa', 'taxa_searched', 'superkingdom', 'kingdom', 'superphylum', 'phylum', 'subphylum', 'superclass', 'class', 'subclass', 'superorder', 'order', 'suborder', 'superfamily', 'family', 'subfamily', 'genus', 'subgenus', 'species')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "fb84a64f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "68565\tuser_taxa\tActinomadura hibisca\n",
      "68565\ttaxa_searched\tActinomadura hibisca\n",
      "68565\tsuperkingdom\tBacteria\t2\n",
      "68565\tkingdom\tNA\tNA\n",
      "68565\tsuperphylum\tNA\tNA\n",
      "68565\tphylum\tActinomycetota\t201174\n",
      "68565\tsubphylum\tNA\tNA\n",
      "68565\tsuperclass\tNA\tNA\n",
      "68565\tclass\tActinomycetes\t1760\n",
      "68565\tsubclass\tNA\tNA\n",
      "68565\tsuperorder\tNA\tNA\n",
      "68565\torder\tStreptosporangiales\t85012\n",
      "68565\tsuborder\tNA\tNA\n",
      "68565\tsuperfamily\tNA\tNA\n",
      "68565\tfamily\tThermomonosporaceae\t2012\n",
      "68565\tsubfamily\tNA\tNA\n",
      "68565\tgenus\tActinomadura\t1988\n",
      "68565\tsubgenus\tNA\tNA\n",
      "68565\tspecies\tActinomadura hibisca\t68565\n"
     ]
    }
   ],
   "source": [
    "for potential_taxid in rank_taxon.lineages:\n",
    "     for rank in ranks:\n",
    "         if rank in ('user_taxa', 'taxa_searched'):\n",
    "             taxon = rank_taxon.lineages[potential_taxid][rank]\n",
    "             print(potential_taxid, rank, taxon, sep='\\t')\n",
    "         else:\n",
    "             taxon, taxid_of_taxon = rank_taxon.lineages[potential_taxid][rank]\n",
    "             print(potential_taxid, rank, taxon, taxid_of_taxon, sep='\\t')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a34ffbd",
   "metadata": {},
   "source": [
    "## Probar helpers para BBH:\n",
    "**<center>1.- best_hit_central_to_orgs:</center>**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "ae41195d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import evo_helpers as hlp\n",
    "import evo_constants as CTS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "a3a1ed0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "blast_File = '/home/luisyovanny/.evo/dev_Evo/data/data_bases/blastp/central_to_genomes/central_to_genomes.blast'\n",
    "rast_Ids = CTS.EXMPL_DB[\"rast_ids\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "286fc82b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Crea el archivo central_to_orgs_bh.uniq:\n",
    "hlp.best_hit_central_to_orgs(blast_File)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "6ba99eb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "hlp.ids_to_names_genomes_file(rast_Ids)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9b3e61d",
   "metadata": {},
   "source": [
    "### best_hit_central_to_genomes está bien!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "f0b88a25",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_central_file_info():\n",
    "    \"\"\"This function creates a file with the names of all\n",
    "    the elements in central_db.\n",
    "    \"\"\"\n",
    "    central_fasta_file_path = CTS.EXMPL_DB[\"central\"] + \"central.fasta\"\n",
    "    output_path = CTS.BBH_aux_files + \"central_enzimes.txt\"\n",
    "    with open(central_fasta_file_path,\"r\") as fasta_file:\n",
    "        with open(output_path, \"w\") as output:\n",
    "            for line in fasta_file:\n",
    "                if \">\" in line:\n",
    "                    Line = line.split(\">\")[1]\n",
    "                    output.write(Line)\n",
    "    return\n",
    "\n",
    "make_central_file_info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "6f93857e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def ids_to_names_genomes_file(file_rast_ids):\n",
    "    \"\"\" This function returns a file with the genome id's as keys\n",
    "    and the genome's names as values. The input file_rast_ids is the Rast.ids\n",
    "    file within the genomes' database.\n",
    "    \"\"\"\n",
    "    \n",
    "    \n",
    "    output_path = CTS.BBH_aux_files + \"ids_to_orgs_names.txt\"\n",
    "    with open(file_rast_ids, \"r\") as file:\n",
    "        with open(output_path, \"w\") as output:\n",
    "            for line in file:\n",
    "                Line = line.split(\"\\t\")[1] + \"-->\" + line.split(\"\\t\")[2]\n",
    "                output.write(Line)\n",
    "            \n",
    "    return \n",
    "rast_ids_File = CTS.EXMPL_DB[\"rast_ids\"]\n",
    "ids_to_names_genomes_file(rast_ids_File)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "603dda8e",
   "metadata": {},
   "source": [
    "# Por hacer:\n",
    "- Agregar una parte en la que se creen todos los archivos auxiliares al comienzo de la rutina, para que estén disponibles en las subrutinas."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1d8632f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "b8d00f70",
   "metadata": {},
   "source": [
    "## Prueba best hit 1 con archivos:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "5a859ed8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def best_hit_central_to_orgs(blast_file):\n",
    "    \"\"\" This function will obtain the best hit from central to organisms. Its output \n",
    "    is a file with these hits. From the output we will obtain the expanded families.\n",
    "    \"\"\"\n",
    "    \n",
    "    # Charge the dataframe with the blastp file:\n",
    "    df = pd.read_csv(blast_file, sep = '\\t', names = CTS.BLAST_COLS)\n",
    "    \n",
    "    #Obtain the different central enzimes:\n",
    "    #central_enzimes = df['query'].unique()\n",
    "    \n",
    "    #Files with the enzimes and orgs info:\n",
    "    central_enzimes_file = CTS.BBH_aux_files + 'central_enzimes.txt'\n",
    "    central_enzimes = []\n",
    "    with open(central_enzimes_file, \"r\") as central_file:\n",
    "        for line in central_file:\n",
    "            if line:\n",
    "                central_enzimes.append(line.rstrip(\"\\n\"))\n",
    "    \n",
    "    ids_to_orgs_names_file = CTS.BBH_aux_files + 'ids_to_orgs_names.txt'\n",
    "    genomes = dict()\n",
    "    with open(ids_to_orgs_names_file, \"r\") as orgs_file:\n",
    "        for line in orgs_file:\n",
    "            if line:\n",
    "                key = line.split(\"-->\")[0]\n",
    "                value = line.split(\"-->\")[1]\n",
    "                genomes[key] = value\n",
    "    \n",
    "    #Obtain a dictionary of the form {genome_id: \"organism_name\"}\n",
    "    #genomes = dict_ids_to_names_genomes(rast_ids)\n",
    "    \n",
    "    #Define and initialize the output dictionary:\n",
    "    unique_central_to_genomes = dict()\n",
    "    for enzime in central_enzimes:\n",
    "        unique_central_to_genomes[enzime] = []\n",
    "        \n",
    "    # Finding the best hit:\n",
    "    for enzime in central_enzimes:\n",
    "        for genome_id in genomes.keys():\n",
    "            aux_list = []\n",
    "            if genome_id != \"6666666.385189\":\n",
    "                for index in range(len(df)):\n",
    "                    if (aux_list == []):\n",
    "                        if ((enzime == df[\"query\"][index])&(genome_id in df[\"subject\"][index])):\n",
    "                            aux_list = list(df.iloc[index])\n",
    "                        else:\n",
    "                            continue\n",
    "                    else:\n",
    "                        if ((enzime == df[\"query\"][index]) & (genome_id in df[\"subject\"][index]) & (df[\"bitscore\"][index] > aux_list[11])):\n",
    "                            aux_list = list(df.iloc[index])\n",
    "                if aux_list != []:  \n",
    "                    unique_central_to_genomes[enzime].append(aux_list)\n",
    "     \n",
    "    return unique_central_to_genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "ad249d44",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "119"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test = best_hit_central_to_orgs(blast_File)\n",
    "tot = 0\n",
    "for enz in test:\n",
    "    tot += len(test[enz])\n",
    "tot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "674fa004",
   "metadata": {},
   "outputs": [],
   "source": [
    "def best_hit_central_to_orgs_2(blast_file):\n",
    "    \"\"\" This function will obtain the best hit from central to organisms. Its output \n",
    "    is a file with these hits. From the output we will obtain the expanded families.\n",
    "    \"\"\"\n",
    "    \n",
    "    # Charge the dataframe with the blastp file:\n",
    "    df = pd.read_csv(blast_file, sep = '\\t', names = CTS.BLAST_COLS)\n",
    "    \n",
    "    ### SE NECESITA QUE LOS ARCHIVOS AUXILIARES ESTÉN CREADOS ###\n",
    "    #Files with the enzimes and orgs info:\n",
    "    \n",
    "    # Make a list with the enzimes:\n",
    "    central_enzimes_file = CTS.BBH_aux_files + 'central_enzimes.txt'\n",
    "    central_enzimes = []\n",
    "    with open(central_enzimes_file, \"r\") as central_file:\n",
    "        for line in central_file:\n",
    "            if line:\n",
    "                central_enzimes.append(line.rstrip(\"\\n\"))\n",
    "    \n",
    "    \n",
    "    # Make a dictionary with the ids and names of organisms:\n",
    "    ids_to_orgs_names_file = CTS.BBH_aux_files + 'ids_to_orgs_names.txt'\n",
    "    genomes = dict()\n",
    "    with open(ids_to_orgs_names_file, \"r\") as orgs_file:\n",
    "        for line in orgs_file:\n",
    "            if line:\n",
    "                key = line.split(\"-->\")[0]\n",
    "                value = line.split(\"-->\")[1]\n",
    "                genomes[key] = value\n",
    "    \n",
    "    \n",
    "    \n",
    "    #Define and initialize the output dictionary:\n",
    "    unique_central_to_genomes = dict()\n",
    "    for enzime in central_enzimes:\n",
    "        unique_central_to_genomes[enzime] = []\n",
    "        \n",
    "    # Finding the best hit:\n",
    "    for enzime in central_enzimes:\n",
    "        for genome_id in genomes.keys():\n",
    "            aux_list = []\n",
    "            \n",
    "            for index in range(len(df)):\n",
    "                if (aux_list == []):\n",
    "                    if ((enzime == df[\"query\"][index])&(genome_id in df[\"subject\"][index])):\n",
    "                        aux_list = list(df.iloc[index])\n",
    "                    else:\n",
    "                        continue\n",
    "                else:\n",
    "                    if ((enzime == df[\"query\"][index]) & (genome_id in df[\"subject\"][index]) & (df[\"bitscore\"][index] > aux_list[11])):\n",
    "                        aux_list = list(df.iloc[index])\n",
    "            if aux_list != []:  \n",
    "                unique_central_to_genomes[enzime].append(aux_list)\n",
    "     \n",
    "    return unique_central_to_genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "2fc065d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_1|Cglu': [['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_1|Cglu',\n",
       "   'gi|1156913.7.6305|1156913.7|NC_021252.1|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Amycolatopsis_orientalis_HCCB10007_NC_021252.1',\n",
       "   59.203,\n",
       "   527,\n",
       "   214,\n",
       "   1,\n",
       "   1,\n",
       "   526,\n",
       "   1,\n",
       "   527,\n",
       "   0.0,\n",
       "   613.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_1|Cglu',\n",
       "   'gi|169133.3.2692|169133.3|LAB|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Citricoccus_sp_CH26_LAB',\n",
       "   46.183,\n",
       "   524,\n",
       "   280,\n",
       "   2,\n",
       "   1,\n",
       "   523,\n",
       "   1,\n",
       "   523,\n",
       "   2.16e-155,\n",
       "   455.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_1|Cglu',\n",
       "   'gi|290340.24.2545|290340.24|NC_008711.1|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Arthrobacter_aurescens_TC1_NC_008711.1',\n",
       "   49.049,\n",
       "   526,\n",
       "   267,\n",
       "   1,\n",
       "   6,\n",
       "   530,\n",
       "   5,\n",
       "   530,\n",
       "   2.0600000000000003e-161,\n",
       "   470.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_1|Cglu',\n",
       "   'gi|367928.21.1188|367928.21|NC_008618.1|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Bifidobacterium_adolescentis_ATCC_15703_NC_008618.1',\n",
       "   34.098,\n",
       "   305,\n",
       "   182,\n",
       "   6,\n",
       "   7,\n",
       "   299,\n",
       "   6,\n",
       "   303,\n",
       "   1.04e-41,\n",
       "   155.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_1|Cglu',\n",
       "   'gi|442563.11.829|442563.11|AD011|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Bifidobacterium_animalis_subsp._lactis_AD011',\n",
       "   36.749,\n",
       "   283,\n",
       "   171,\n",
       "   3,\n",
       "   15,\n",
       "   292,\n",
       "   13,\n",
       "   292,\n",
       "   3.03e-46,\n",
       "   167.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_1|Cglu',\n",
       "   'gi|446462.15.6083|446462.15|NC_021252.1|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Actinosynnema_mirum_DSM_43827_NC_013093.1_NC_021252.1',\n",
       "   60.338,\n",
       "   532,\n",
       "   209,\n",
       "   2,\n",
       "   1,\n",
       "   530,\n",
       "   1,\n",
       "   532,\n",
       "   0.0,\n",
       "   595.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_1|Cglu',\n",
       "   'gi|452863.24.2320|452863.24|NC_011886.1|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Arthrobacter_chlorophenolicus_A6_NC_011886.1',\n",
       "   49.43,\n",
       "   526,\n",
       "   265,\n",
       "   1,\n",
       "   6,\n",
       "   530,\n",
       "   3,\n",
       "   528,\n",
       "   7.31e-164,\n",
       "   476.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_1|Cglu',\n",
       "   'gi|649831.6.6869|649831.6|NC_021191.1|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Actinoplanes_sp._N902-109_NC_021191.1',\n",
       "   52.408,\n",
       "   519,\n",
       "   244,\n",
       "   2,\n",
       "   7,\n",
       "   522,\n",
       "   3,\n",
       "   521,\n",
       "   6.7300000000000004e-167,\n",
       "   484.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_1|Cglu',\n",
       "   'gi|6666666.104540.4265|6666666.104540|AOHO01|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Amycolatopsis_decaplanina_DSM_44594_AOHO01',\n",
       "   59.203,\n",
       "   527,\n",
       "   214,\n",
       "   1,\n",
       "   1,\n",
       "   526,\n",
       "   1,\n",
       "   527,\n",
       "   0.0,\n",
       "   612.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_1|Cglu',\n",
       "   'gi|6666666.111158.7934|6666666.111158|NC_0131311|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Catenulispora_acidiphila_DSM_44928_NC_0131311',\n",
       "   54.685,\n",
       "   523,\n",
       "   237,\n",
       "   0,\n",
       "   4,\n",
       "   526,\n",
       "   7,\n",
       "   529,\n",
       "   3.6999999999999997e-180,\n",
       "   518.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_1|Cglu',\n",
       "   'gi|6666666.111572.1121|6666666.111572|NC_0094801|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Clavibacter_michiganensis_subsp._michiganensis_NCPPB_382_NC_0094801',\n",
       "   49.904,\n",
       "   521,\n",
       "   261,\n",
       "   0,\n",
       "   6,\n",
       "   526,\n",
       "   3,\n",
       "   523,\n",
       "   1.55e-162,\n",
       "   473.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_1|Cglu',\n",
       "   'gi|6666666.112811.5461|6666666.112811|JOEF01|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Allokutzneria_albata_JOEF01',\n",
       "   58.317,\n",
       "   523,\n",
       "   217,\n",
       "   1,\n",
       "   1,\n",
       "   522,\n",
       "   1,\n",
       "   523,\n",
       "   0.0,\n",
       "   580.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_1|Cglu',\n",
       "   'gi|6666666.146813.1460|6666666.146813|NC_014550.1|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Arthrobacter_arilaitensis_RE117_NC_014550.1',\n",
       "   49.522,\n",
       "   523,\n",
       "   263,\n",
       "   1,\n",
       "   1,\n",
       "   523,\n",
       "   1,\n",
       "   522,\n",
       "   3.7e-170,\n",
       "   492.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_1|Cglu',\n",
       "   'gi|6666666.146852.1090|6666666.146852|NC_002935.2|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Corynebacterium_diphtheriae_NCTC_13129_NC_002935.2',\n",
       "   79.623,\n",
       "   530,\n",
       "   108,\n",
       "   0,\n",
       "   1,\n",
       "   530,\n",
       "   1,\n",
       "   530,\n",
       "   0.0,\n",
       "   847.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_1|Cglu',\n",
       "   'gi|6666666.146872.7282|6666666.146872|NC_017803.1|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Actinoplanes_sp._SE50/110_NC_017803.1',\n",
       "   51.145,\n",
       "   524,\n",
       "   248,\n",
       "   3,\n",
       "   8,\n",
       "   529,\n",
       "   4,\n",
       "   521,\n",
       "   6.18e-171,\n",
       "   494.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_1|Cglu',\n",
       "   'gi|749927.13.1727|749927.13|NC_014318.1|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Amycolatopsis_mediterranei_U32_NC_014318.1',\n",
       "   59.586,\n",
       "   532,\n",
       "   213,\n",
       "   2,\n",
       "   1,\n",
       "   530,\n",
       "   1,\n",
       "   532,\n",
       "   0.0,\n",
       "   620.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_1|Cglu',\n",
       "   'gi|930171.11.2378|930171.11|NC_015145.1|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Arthrobacter_phenanthrenivorans_Sphe3_NC_015145.1',\n",
       "   49.43,\n",
       "   526,\n",
       "   265,\n",
       "   1,\n",
       "   6,\n",
       "   530,\n",
       "   3,\n",
       "   528,\n",
       "   4.469999999999999e-165,\n",
       "   479.0]],\n",
       " '3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_2|Mtub': [['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_2|Mtub',\n",
       "   'gi|1156913.7.7552|1156913.7|NC_021252.1|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Amycolatopsis_orientalis_HCCB10007_NC_021252.1',\n",
       "   40.769,\n",
       "   260,\n",
       "   144,\n",
       "   3,\n",
       "   73,\n",
       "   324,\n",
       "   85,\n",
       "   342,\n",
       "   6.64e-47,\n",
       "   162.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_2|Mtub',\n",
       "   'gi|169133.3.2692|169133.3|LAB|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Citricoccus_sp_CH26_LAB',\n",
       "   35.294,\n",
       "   255,\n",
       "   150,\n",
       "   3,\n",
       "   74,\n",
       "   320,\n",
       "   68,\n",
       "   315,\n",
       "   1.8e-37,\n",
       "   141.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_2|Mtub',\n",
       "   'gi|290340.24.1979|290340.24|NC_008711.1|Glyoxylate_reductase_(EC_1.1.1.79)_/_Glyoxylate_reductase_(EC_1.1.1.26)_/_Hydroxypyruvate_reductase_(EC_1.1.1.81)|Arthrobacter_aurescens_TC1_NC_008711.1',\n",
       "   34.154,\n",
       "   325,\n",
       "   186,\n",
       "   9,\n",
       "   6,\n",
       "   315,\n",
       "   3,\n",
       "   314,\n",
       "   4.04e-37,\n",
       "   135.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_2|Mtub',\n",
       "   'gi|367928.21.1188|367928.21|NC_008618.1|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Bifidobacterium_adolescentis_ATCC_15703_NC_008618.1',\n",
       "   29.644,\n",
       "   253,\n",
       "   145,\n",
       "   7,\n",
       "   85,\n",
       "   315,\n",
       "   81,\n",
       "   322,\n",
       "   2.67e-23,\n",
       "   100.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_2|Mtub',\n",
       "   'gi|442563.11.379|442563.11|AD011|Phosphoglycerate_dehydrogenase|Bifidobacterium_animalis_subsp._lactis_AD011',\n",
       "   34.583,\n",
       "   240,\n",
       "   135,\n",
       "   5,\n",
       "   82,\n",
       "   311,\n",
       "   100,\n",
       "   327,\n",
       "   4.33e-28,\n",
       "   112.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_2|Mtub',\n",
       "   'gi|446462.15.4214|446462.15|NC_021252.1|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Actinosynnema_mirum_DSM_43827_NC_013093.1_NC_021252.1',\n",
       "   40.288,\n",
       "   278,\n",
       "   149,\n",
       "   5,\n",
       "   39,\n",
       "   304,\n",
       "   33,\n",
       "   305,\n",
       "   3.35e-41,\n",
       "   146.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_2|Mtub',\n",
       "   'gi|452863.24.89|452863.24|NC_011886.1|Glyoxylate_reductase_(EC_1.1.1.79)_/_Glyoxylate_reductase_(EC_1.1.1.26)_/_Hydroxypyruvate_reductase_(EC_1.1.1.81)|Arthrobacter_chlorophenolicus_A6_NC_011886.1',\n",
       "   35.276,\n",
       "   326,\n",
       "   183,\n",
       "   9,\n",
       "   6,\n",
       "   316,\n",
       "   3,\n",
       "   315,\n",
       "   1.24e-40,\n",
       "   145.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_2|Mtub',\n",
       "   'gi|649831.6.6869|649831.6|NC_021191.1|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Actinoplanes_sp._N902-109_NC_021191.1',\n",
       "   38.132,\n",
       "   257,\n",
       "   144,\n",
       "   3,\n",
       "   74,\n",
       "   322,\n",
       "   65,\n",
       "   314,\n",
       "   9.169999999999999e-41,\n",
       "   150.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_2|Mtub',\n",
       "   'gi|6666666.104540.4265|6666666.104540|AOHO01|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Amycolatopsis_decaplanina_DSM_44594_AOHO01',\n",
       "   38.039,\n",
       "   255,\n",
       "   143,\n",
       "   3,\n",
       "   74,\n",
       "   320,\n",
       "   69,\n",
       "   316,\n",
       "   1.46e-39,\n",
       "   146.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_2|Mtub',\n",
       "   'gi|6666666.111158.7934|6666666.111158|NC_0131311|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Catenulispora_acidiphila_DSM_44928_NC_0131311',\n",
       "   36.471,\n",
       "   255,\n",
       "   147,\n",
       "   4,\n",
       "   74,\n",
       "   320,\n",
       "   72,\n",
       "   319,\n",
       "   6.21e-33,\n",
       "   128.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_2|Mtub',\n",
       "   'gi|6666666.111572.1121|6666666.111572|NC_0094801|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Clavibacter_michiganensis_subsp._michiganensis_NCPPB_382_NC_0094801',\n",
       "   35.686,\n",
       "   255,\n",
       "   149,\n",
       "   3,\n",
       "   74,\n",
       "   320,\n",
       "   66,\n",
       "   313,\n",
       "   7.95e-37,\n",
       "   139.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_2|Mtub',\n",
       "   'gi|6666666.112811.1464|6666666.112811|JOEF01|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Allokutzneria_albata_JOEF01',\n",
       "   41.606,\n",
       "   274,\n",
       "   152,\n",
       "   4,\n",
       "   56,\n",
       "   323,\n",
       "   67,\n",
       "   338,\n",
       "   4.08e-56,\n",
       "   186.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_2|Mtub',\n",
       "   'gi|6666666.146813.1460|6666666.146813|NC_014550.1|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Arthrobacter_arilaitensis_RE117_NC_014550.1',\n",
       "   36.187,\n",
       "   257,\n",
       "   149,\n",
       "   3,\n",
       "   72,\n",
       "   320,\n",
       "   66,\n",
       "   315,\n",
       "   1.8e-41,\n",
       "   152.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_2|Mtub',\n",
       "   'gi|6666666.146852.1090|6666666.146852|NC_002935.2|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Corynebacterium_diphtheriae_NCTC_13129_NC_002935.2',\n",
       "   33.548,\n",
       "   310,\n",
       "   187,\n",
       "   6,\n",
       "   23,\n",
       "   320,\n",
       "   14,\n",
       "   316,\n",
       "   1.88e-35,\n",
       "   135.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_2|Mtub',\n",
       "   'gi|6666666.146872.7282|6666666.146872|NC_017803.1|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Actinoplanes_sp._SE50/110_NC_017803.1',\n",
       "   37.956,\n",
       "   274,\n",
       "   153,\n",
       "   4,\n",
       "   57,\n",
       "   320,\n",
       "   46,\n",
       "   312,\n",
       "   7.03e-40,\n",
       "   147.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_2|Mtub',\n",
       "   'gi|749927.13.1727|749927.13|NC_014318.1|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Amycolatopsis_mediterranei_U32_NC_014318.1',\n",
       "   35.691,\n",
       "   311,\n",
       "   181,\n",
       "   6,\n",
       "   22,\n",
       "   320,\n",
       "   13,\n",
       "   316,\n",
       "   1.05e-37,\n",
       "   141.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_2|Mtub',\n",
       "   'gi|930171.11.2378|930171.11|NC_015145.1|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Arthrobacter_phenanthrenivorans_Sphe3_NC_015145.1',\n",
       "   36.187,\n",
       "   257,\n",
       "   149,\n",
       "   3,\n",
       "   72,\n",
       "   320,\n",
       "   64,\n",
       "   313,\n",
       "   2.3700000000000002e-35,\n",
       "   135.0]],\n",
       " '3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_3|Mtub': [['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_3|Mtub',\n",
       "   'gi|1156913.7.6305|1156913.7|NC_021252.1|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Amycolatopsis_orientalis_HCCB10007_NC_021252.1',\n",
       "   64.381,\n",
       "   525,\n",
       "   186,\n",
       "   1,\n",
       "   4,\n",
       "   527,\n",
       "   7,\n",
       "   531,\n",
       "   0.0,\n",
       "   645.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_3|Mtub',\n",
       "   'gi|169133.3.2692|169133.3|LAB|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Citricoccus_sp_CH26_LAB',\n",
       "   50.38,\n",
       "   526,\n",
       "   258,\n",
       "   2,\n",
       "   4,\n",
       "   527,\n",
       "   6,\n",
       "   530,\n",
       "   4.01e-176,\n",
       "   508.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_3|Mtub',\n",
       "   'gi|290340.24.2545|290340.24|NC_008711.1|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Arthrobacter_aurescens_TC1_NC_008711.1',\n",
       "   49.43,\n",
       "   526,\n",
       "   266,\n",
       "   0,\n",
       "   2,\n",
       "   527,\n",
       "   4,\n",
       "   529,\n",
       "   9.35e-171,\n",
       "   494.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_3|Mtub',\n",
       "   'gi|367928.21.1188|367928.21|NC_008618.1|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Bifidobacterium_adolescentis_ATCC_15703_NC_008618.1',\n",
       "   30.492,\n",
       "   305,\n",
       "   203,\n",
       "   5,\n",
       "   2,\n",
       "   299,\n",
       "   4,\n",
       "   306,\n",
       "   1.42e-38,\n",
       "   146.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_3|Mtub',\n",
       "   'gi|442563.11.829|442563.11|AD011|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Bifidobacterium_animalis_subsp._lactis_AD011',\n",
       "   31.353,\n",
       "   303,\n",
       "   197,\n",
       "   4,\n",
       "   3,\n",
       "   297,\n",
       "   1,\n",
       "   300,\n",
       "   5.5599999999999995e-40,\n",
       "   150.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_3|Mtub',\n",
       "   'gi|446462.15.6083|446462.15|NC_021252.1|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Actinosynnema_mirum_DSM_43827_NC_013093.1_NC_021252.1',\n",
       "   65.465,\n",
       "   527,\n",
       "   181,\n",
       "   1,\n",
       "   2,\n",
       "   527,\n",
       "   5,\n",
       "   531,\n",
       "   0.0,\n",
       "   622.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_3|Mtub',\n",
       "   'gi|452863.24.2320|452863.24|NC_011886.1|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Arthrobacter_chlorophenolicus_A6_NC_011886.1',\n",
       "   49.905,\n",
       "   527,\n",
       "   264,\n",
       "   0,\n",
       "   1,\n",
       "   527,\n",
       "   1,\n",
       "   527,\n",
       "   4.0400000000000006e-173,\n",
       "   499.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_3|Mtub',\n",
       "   'gi|649831.6.6869|649831.6|NC_021191.1|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Actinoplanes_sp._N902-109_NC_021191.1',\n",
       "   55.787,\n",
       "   527,\n",
       "   230,\n",
       "   2,\n",
       "   4,\n",
       "   527,\n",
       "   3,\n",
       "   529,\n",
       "   8.28e-178,\n",
       "   511.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_3|Mtub',\n",
       "   'gi|6666666.104540.4265|6666666.104540|AOHO01|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Amycolatopsis_decaplanina_DSM_44594_AOHO01',\n",
       "   64.381,\n",
       "   525,\n",
       "   186,\n",
       "   1,\n",
       "   4,\n",
       "   527,\n",
       "   7,\n",
       "   531,\n",
       "   0.0,\n",
       "   643.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_3|Mtub',\n",
       "   'gi|6666666.111158.7934|6666666.111158|NC_0131311|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Catenulispora_acidiphila_DSM_44928_NC_0131311',\n",
       "   58.969,\n",
       "   524,\n",
       "   215,\n",
       "   0,\n",
       "   4,\n",
       "   527,\n",
       "   10,\n",
       "   533,\n",
       "   0.0,\n",
       "   566.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_3|Mtub',\n",
       "   'gi|6666666.111572.1121|6666666.111572|NC_0094801|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Clavibacter_michiganensis_subsp._michiganensis_NCPPB_382_NC_0094801',\n",
       "   52.273,\n",
       "   528,\n",
       "   252,\n",
       "   0,\n",
       "   1,\n",
       "   528,\n",
       "   1,\n",
       "   528,\n",
       "   1.17e-175,\n",
       "   506.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_3|Mtub',\n",
       "   'gi|6666666.112811.5461|6666666.112811|JOEF01|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Allokutzneria_albata_JOEF01',\n",
       "   66.151,\n",
       "   517,\n",
       "   174,\n",
       "   1,\n",
       "   4,\n",
       "   519,\n",
       "   7,\n",
       "   523,\n",
       "   0.0,\n",
       "   637.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_3|Mtub',\n",
       "   'gi|6666666.146813.1460|6666666.146813|NC_014550.1|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Arthrobacter_arilaitensis_RE117_NC_014550.1',\n",
       "   51.938,\n",
       "   516,\n",
       "   248,\n",
       "   0,\n",
       "   4,\n",
       "   519,\n",
       "   6,\n",
       "   521,\n",
       "   0.0,\n",
       "   534.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_3|Mtub',\n",
       "   'gi|6666666.146852.1090|6666666.146852|NC_002935.2|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Corynebacterium_diphtheriae_NCTC_13129_NC_002935.2',\n",
       "   67.562,\n",
       "   521,\n",
       "   169,\n",
       "   0,\n",
       "   4,\n",
       "   524,\n",
       "   7,\n",
       "   527,\n",
       "   0.0,\n",
       "   688.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_3|Mtub',\n",
       "   'gi|6666666.146872.7282|6666666.146872|NC_017803.1|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Actinoplanes_sp._SE50/110_NC_017803.1',\n",
       "   55.725,\n",
       "   524,\n",
       "   225,\n",
       "   2,\n",
       "   5,\n",
       "   527,\n",
       "   4,\n",
       "   521,\n",
       "   0.0,\n",
       "   538.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_3|Mtub',\n",
       "   'gi|749927.13.1727|749927.13|NC_014318.1|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Amycolatopsis_mediterranei_U32_NC_014318.1',\n",
       "   65.341,\n",
       "   528,\n",
       "   182,\n",
       "   1,\n",
       "   2,\n",
       "   528,\n",
       "   5,\n",
       "   532,\n",
       "   0.0,\n",
       "   658.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_3|Mtub',\n",
       "   'gi|930171.11.2378|930171.11|NC_015145.1|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Arthrobacter_phenanthrenivorans_Sphe3_NC_015145.1',\n",
       "   50.854,\n",
       "   527,\n",
       "   259,\n",
       "   0,\n",
       "   1,\n",
       "   527,\n",
       "   1,\n",
       "   527,\n",
       "   1.0999999999999999e-174,\n",
       "   504.0]],\n",
       " '3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_4|Scoe': [['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_4|Scoe',\n",
       "   'gi|1156913.7.6305|1156913.7|NC_021252.1|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Amycolatopsis_orientalis_HCCB10007_NC_021252.1',\n",
       "   53.7,\n",
       "   527,\n",
       "   243,\n",
       "   1,\n",
       "   3,\n",
       "   528,\n",
       "   5,\n",
       "   531,\n",
       "   0.0,\n",
       "   545.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_4|Scoe',\n",
       "   'gi|169133.3.2692|169133.3|LAB|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Citricoccus_sp_CH26_LAB',\n",
       "   62.879,\n",
       "   528,\n",
       "   195,\n",
       "   1,\n",
       "   3,\n",
       "   529,\n",
       "   4,\n",
       "   531,\n",
       "   0.0,\n",
       "   686.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_4|Scoe',\n",
       "   'gi|290340.24.2545|290340.24|NC_008711.1|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Arthrobacter_aurescens_TC1_NC_008711.1',\n",
       "   67.362,\n",
       "   527,\n",
       "   172,\n",
       "   0,\n",
       "   2,\n",
       "   528,\n",
       "   3,\n",
       "   529,\n",
       "   0.0,\n",
       "   715.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_4|Scoe',\n",
       "   'gi|367928.21.1188|367928.21|NC_008618.1|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Bifidobacterium_adolescentis_ATCC_15703_NC_008618.1',\n",
       "   32.459,\n",
       "   305,\n",
       "   197,\n",
       "   4,\n",
       "   3,\n",
       "   300,\n",
       "   4,\n",
       "   306,\n",
       "   6.57e-43,\n",
       "   158.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_4|Scoe',\n",
       "   'gi|442563.11.829|442563.11|AD011|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Bifidobacterium_animalis_subsp._lactis_AD011',\n",
       "   33.77,\n",
       "   305,\n",
       "   191,\n",
       "   5,\n",
       "   5,\n",
       "   302,\n",
       "   2,\n",
       "   302,\n",
       "   3.6e-43,\n",
       "   159.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_4|Scoe',\n",
       "   'gi|446462.15.6083|446462.15|NC_021252.1|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Actinosynnema_mirum_DSM_43827_NC_013093.1_NC_021252.1',\n",
       "   56.061,\n",
       "   528,\n",
       "   231,\n",
       "   1,\n",
       "   2,\n",
       "   528,\n",
       "   4,\n",
       "   531,\n",
       "   0.0,\n",
       "   541.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_4|Scoe',\n",
       "   'gi|452863.24.2320|452863.24|NC_011886.1|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Arthrobacter_chlorophenolicus_A6_NC_011886.1',\n",
       "   65.779,\n",
       "   526,\n",
       "   180,\n",
       "   0,\n",
       "   3,\n",
       "   528,\n",
       "   2,\n",
       "   527,\n",
       "   0.0,\n",
       "   700.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_4|Scoe',\n",
       "   'gi|649831.6.6869|649831.6|NC_021191.1|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Actinoplanes_sp._N902-109_NC_021191.1',\n",
       "   61.48,\n",
       "   527,\n",
       "   200,\n",
       "   2,\n",
       "   5,\n",
       "   528,\n",
       "   3,\n",
       "   529,\n",
       "   0.0,\n",
       "   606.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_4|Scoe',\n",
       "   'gi|6666666.104540.4265|6666666.104540|AOHO01|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Amycolatopsis_decaplanina_DSM_44594_AOHO01',\n",
       "   53.89,\n",
       "   527,\n",
       "   242,\n",
       "   1,\n",
       "   3,\n",
       "   528,\n",
       "   5,\n",
       "   531,\n",
       "   0.0,\n",
       "   545.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_4|Scoe',\n",
       "   'gi|6666666.111158.7934|6666666.111158|NC_0131311|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Catenulispora_acidiphila_DSM_44928_NC_0131311',\n",
       "   74.053,\n",
       "   528,\n",
       "   137,\n",
       "   0,\n",
       "   1,\n",
       "   528,\n",
       "   6,\n",
       "   533,\n",
       "   0.0,\n",
       "   769.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_4|Scoe',\n",
       "   'gi|6666666.111572.1121|6666666.111572|NC_0094801|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Clavibacter_michiganensis_subsp._michiganensis_NCPPB_382_NC_0094801',\n",
       "   64.259,\n",
       "   526,\n",
       "   188,\n",
       "   0,\n",
       "   3,\n",
       "   528,\n",
       "   2,\n",
       "   527,\n",
       "   0.0,\n",
       "   672.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_4|Scoe',\n",
       "   'gi|6666666.112811.5461|6666666.112811|JOEF01|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Allokutzneria_albata_JOEF01',\n",
       "   55.682,\n",
       "   528,\n",
       "   233,\n",
       "   1,\n",
       "   2,\n",
       "   528,\n",
       "   4,\n",
       "   531,\n",
       "   0.0,\n",
       "   563.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_4|Scoe',\n",
       "   'gi|6666666.146813.1460|6666666.146813|NC_014550.1|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Arthrobacter_arilaitensis_RE117_NC_014550.1',\n",
       "   65.779,\n",
       "   526,\n",
       "   180,\n",
       "   0,\n",
       "   3,\n",
       "   528,\n",
       "   4,\n",
       "   529,\n",
       "   0.0,\n",
       "   718.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_4|Scoe',\n",
       "   'gi|6666666.146852.1090|6666666.146852|NC_002935.2|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Corynebacterium_diphtheriae_NCTC_13129_NC_002935.2',\n",
       "   51.731,\n",
       "   520,\n",
       "   251,\n",
       "   0,\n",
       "   1,\n",
       "   520,\n",
       "   3,\n",
       "   522,\n",
       "   6.23e-177,\n",
       "   509.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_4|Scoe',\n",
       "   'gi|6666666.146872.7282|6666666.146872|NC_017803.1|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Actinoplanes_sp._SE50/110_NC_017803.1',\n",
       "   60.496,\n",
       "   524,\n",
       "   200,\n",
       "   3,\n",
       "   6,\n",
       "   528,\n",
       "   4,\n",
       "   521,\n",
       "   0.0,\n",
       "   612.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_4|Scoe',\n",
       "   'gi|749927.13.1727|749927.13|NC_014318.1|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Amycolatopsis_mediterranei_U32_NC_014318.1',\n",
       "   53.321,\n",
       "   527,\n",
       "   245,\n",
       "   1,\n",
       "   3,\n",
       "   528,\n",
       "   5,\n",
       "   531,\n",
       "   0.0,\n",
       "   546.0],\n",
       "  ['3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_4|Scoe',\n",
       "   'gi|930171.11.2378|930171.11|NC_015145.1|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Arthrobacter_phenanthrenivorans_Sphe3_NC_015145.1',\n",
       "   66.16,\n",
       "   526,\n",
       "   178,\n",
       "   0,\n",
       "   3,\n",
       "   528,\n",
       "   2,\n",
       "   527,\n",
       "   0.0,\n",
       "   702.0]],\n",
       " '3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_1|none': [['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_1|none',\n",
       "   'gi|1156913.7.7342|1156913.7|NC_021252.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Amycolatopsis_orientalis_HCCB10007_NC_021252.1',\n",
       "   62.634,\n",
       "   372,\n",
       "   137,\n",
       "   1,\n",
       "   1,\n",
       "   370,\n",
       "   1,\n",
       "   372,\n",
       "   1.49e-168,\n",
       "   476.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_1|none',\n",
       "   'gi|169133.3.2262|169133.3|LAB|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Citricoccus_sp_CH26_LAB',\n",
       "   61.202,\n",
       "   366,\n",
       "   141,\n",
       "   1,\n",
       "   6,\n",
       "   370,\n",
       "   7,\n",
       "   372,\n",
       "   3.75e-167,\n",
       "   473.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_1|none',\n",
       "   'gi|290340.24.1033|290340.24|NC_008711.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Arthrobacter_aurescens_TC1_NC_008711.1',\n",
       "   62.466,\n",
       "   365,\n",
       "   136,\n",
       "   1,\n",
       "   6,\n",
       "   369,\n",
       "   7,\n",
       "   371,\n",
       "   3.56e-169,\n",
       "   478.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_1|none',\n",
       "   'gi|367928.21.407|367928.21|NC_008618.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Bifidobacterium_adolescentis_ATCC_15703_NC_008618.1',\n",
       "   54.233,\n",
       "   378,\n",
       "   165,\n",
       "   2,\n",
       "   1,\n",
       "   370,\n",
       "   1,\n",
       "   378,\n",
       "   2.83e-146,\n",
       "   420.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_1|none',\n",
       "   'gi|442563.11.458|442563.11|AD011|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Bifidobacterium_animalis_subsp._lactis_AD011',\n",
       "   49.598,\n",
       "   373,\n",
       "   180,\n",
       "   2,\n",
       "   6,\n",
       "   370,\n",
       "   5,\n",
       "   377,\n",
       "   2.18e-132,\n",
       "   384.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_1|none',\n",
       "   'gi|446462.15.520|446462.15|NC_021252.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Actinosynnema_mirum_DSM_43827_NC_013093.1_NC_021252.1',\n",
       "   63.636,\n",
       "   363,\n",
       "   132,\n",
       "   0,\n",
       "   7,\n",
       "   369,\n",
       "   12,\n",
       "   374,\n",
       "   4.53e-176,\n",
       "   495.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_1|none',\n",
       "   'gi|452863.24.923|452863.24|NC_011886.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Arthrobacter_chlorophenolicus_A6_NC_011886.1',\n",
       "   61.096,\n",
       "   365,\n",
       "   141,\n",
       "   1,\n",
       "   6,\n",
       "   369,\n",
       "   7,\n",
       "   371,\n",
       "   1.06e-164,\n",
       "   466.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_1|none',\n",
       "   'gi|649831.6.513|649831.6|NC_021191.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Actinoplanes_sp._N902-109_NC_021191.1',\n",
       "   53.639,\n",
       "   371,\n",
       "   171,\n",
       "   1,\n",
       "   1,\n",
       "   370,\n",
       "   1,\n",
       "   371,\n",
       "   7.34e-140,\n",
       "   403.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_1|none',\n",
       "   'gi|6666666.104540.2193|6666666.104540|AOHO01|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Amycolatopsis_decaplanina_DSM_44594_AOHO01',\n",
       "   62.634,\n",
       "   372,\n",
       "   137,\n",
       "   1,\n",
       "   1,\n",
       "   370,\n",
       "   1,\n",
       "   372,\n",
       "   1.37e-174,\n",
       "   491.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_1|none',\n",
       "   'gi|6666666.111158.8940|6666666.111158|NC_0131311|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Catenulispora_acidiphila_DSM_44928_NC_0131311',\n",
       "   60.109,\n",
       "   366,\n",
       "   145,\n",
       "   1,\n",
       "   6,\n",
       "   370,\n",
       "   5,\n",
       "   370,\n",
       "   2.3900000000000003e-164,\n",
       "   465.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_1|none',\n",
       "   'gi|6666666.111572.2478|6666666.111572|NC_0094801|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Clavibacter_michiganensis_subsp._michiganensis_NCPPB_382_NC_0094801',\n",
       "   59.401,\n",
       "   367,\n",
       "   148,\n",
       "   1,\n",
       "   6,\n",
       "   372,\n",
       "   297,\n",
       "   662,\n",
       "   2.37e-147,\n",
       "   432.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_1|none',\n",
       "   'gi|6666666.112811.7991|6666666.112811|JOEF01|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Allokutzneria_albata_JOEF01',\n",
       "   64.011,\n",
       "   364,\n",
       "   131,\n",
       "   0,\n",
       "   7,\n",
       "   370,\n",
       "   12,\n",
       "   375,\n",
       "   3.65e-176,\n",
       "   495.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_1|none',\n",
       "   'gi|6666666.146813.572|6666666.146813|NC_014550.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Arthrobacter_arilaitensis_RE117_NC_014550.1',\n",
       "   59.452,\n",
       "   365,\n",
       "   147,\n",
       "   1,\n",
       "   7,\n",
       "   370,\n",
       "   7,\n",
       "   371,\n",
       "   6.460000000000001e-161,\n",
       "   456.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_1|none',\n",
       "   'gi|6666666.146852.774|6666666.146852|NC_002935.2|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Corynebacterium_diphtheriae_NCTC_13129_NC_002935.2',\n",
       "   81.6,\n",
       "   375,\n",
       "   69,\n",
       "   0,\n",
       "   1,\n",
       "   375,\n",
       "   1,\n",
       "   375,\n",
       "   0.0,\n",
       "   646.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_1|none',\n",
       "   'gi|6666666.146872.438|6666666.146872|NC_017803.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Actinoplanes_sp._SE50/110_NC_017803.1',\n",
       "   52.603,\n",
       "   365,\n",
       "   172,\n",
       "   1,\n",
       "   7,\n",
       "   370,\n",
       "   8,\n",
       "   372,\n",
       "   5.71e-137,\n",
       "   396.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_1|none',\n",
       "   'gi|749927.13.8695|749927.13|NC_014318.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Amycolatopsis_mediterranei_U32_NC_014318.1',\n",
       "   62.534,\n",
       "   371,\n",
       "   138,\n",
       "   1,\n",
       "   1,\n",
       "   370,\n",
       "   1,\n",
       "   371,\n",
       "   7.59e-172,\n",
       "   484.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_1|none',\n",
       "   'gi|930171.11.818|930171.11|NC_015145.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Arthrobacter_phenanthrenivorans_Sphe3_NC_015145.1',\n",
       "   61.096,\n",
       "   365,\n",
       "   141,\n",
       "   1,\n",
       "   6,\n",
       "   369,\n",
       "   7,\n",
       "   371,\n",
       "   1.55e-165,\n",
       "   468.0]],\n",
       " '3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_2|none': [['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_2|none',\n",
       "   'gi|1156913.7.7342|1156913.7|NC_021252.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Amycolatopsis_orientalis_HCCB10007_NC_021252.1',\n",
       "   67.027,\n",
       "   370,\n",
       "   122,\n",
       "   0,\n",
       "   7,\n",
       "   376,\n",
       "   5,\n",
       "   374,\n",
       "   0.0,\n",
       "   516.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_2|none',\n",
       "   'gi|169133.3.2262|169133.3|LAB|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Citricoccus_sp_CH26_LAB',\n",
       "   64.785,\n",
       "   372,\n",
       "   130,\n",
       "   1,\n",
       "   6,\n",
       "   376,\n",
       "   3,\n",
       "   374,\n",
       "   1.2e-179,\n",
       "   504.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_2|none',\n",
       "   'gi|290340.24.1033|290340.24|NC_008711.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Arthrobacter_aurescens_TC1_NC_008711.1',\n",
       "   60.976,\n",
       "   369,\n",
       "   143,\n",
       "   1,\n",
       "   9,\n",
       "   376,\n",
       "   6,\n",
       "   374,\n",
       "   3.3e-169,\n",
       "   478.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_2|none',\n",
       "   'gi|367928.21.407|367928.21|NC_008618.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Bifidobacterium_adolescentis_ATCC_15703_NC_008618.1',\n",
       "   56.053,\n",
       "   380,\n",
       "   159,\n",
       "   2,\n",
       "   5,\n",
       "   376,\n",
       "   1,\n",
       "   380,\n",
       "   2.3099999999999998e-150,\n",
       "   430.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_2|none',\n",
       "   'gi|442563.11.458|442563.11|AD011|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Bifidobacterium_animalis_subsp._lactis_AD011',\n",
       "   54.521,\n",
       "   376,\n",
       "   163,\n",
       "   2,\n",
       "   9,\n",
       "   376,\n",
       "   4,\n",
       "   379,\n",
       "   3.52e-144,\n",
       "   414.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_2|none',\n",
       "   'gi|446462.15.520|446462.15|NC_021252.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Actinosynnema_mirum_DSM_43827_NC_013093.1_NC_021252.1',\n",
       "   70.109,\n",
       "   368,\n",
       "   110,\n",
       "   0,\n",
       "   9,\n",
       "   376,\n",
       "   10,\n",
       "   377,\n",
       "   0.0,\n",
       "   537.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_2|none',\n",
       "   'gi|452863.24.923|452863.24|NC_011886.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Arthrobacter_chlorophenolicus_A6_NC_011886.1',\n",
       "   59.892,\n",
       "   369,\n",
       "   147,\n",
       "   1,\n",
       "   9,\n",
       "   376,\n",
       "   6,\n",
       "   374,\n",
       "   1.6800000000000002e-164,\n",
       "   466.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_2|none',\n",
       "   'gi|649831.6.513|649831.6|NC_021191.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Actinoplanes_sp._N902-109_NC_021191.1',\n",
       "   57.995,\n",
       "   369,\n",
       "   154,\n",
       "   1,\n",
       "   9,\n",
       "   376,\n",
       "   5,\n",
       "   373,\n",
       "   1.99e-152,\n",
       "   435.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_2|none',\n",
       "   'gi|6666666.104540.2193|6666666.104540|AOHO01|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Amycolatopsis_decaplanina_DSM_44594_AOHO01',\n",
       "   67.027,\n",
       "   370,\n",
       "   122,\n",
       "   0,\n",
       "   7,\n",
       "   376,\n",
       "   5,\n",
       "   374,\n",
       "   0.0,\n",
       "   532.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_2|none',\n",
       "   'gi|6666666.111158.8940|6666666.111158|NC_0131311|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Catenulispora_acidiphila_DSM_44928_NC_0131311',\n",
       "   63.686,\n",
       "   369,\n",
       "   133,\n",
       "   1,\n",
       "   9,\n",
       "   376,\n",
       "   4,\n",
       "   372,\n",
       "   8.41e-175,\n",
       "   492.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_2|none',\n",
       "   'gi|6666666.111572.2478|6666666.111572|NC_0094801|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Clavibacter_michiganensis_subsp._michiganensis_NCPPB_382_NC_0094801',\n",
       "   60.976,\n",
       "   369,\n",
       "   143,\n",
       "   1,\n",
       "   7,\n",
       "   375,\n",
       "   294,\n",
       "   661,\n",
       "   1.14e-153,\n",
       "   449.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_2|none',\n",
       "   'gi|6666666.112811.7991|6666666.112811|JOEF01|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Allokutzneria_albata_JOEF01',\n",
       "   69.022,\n",
       "   368,\n",
       "   114,\n",
       "   0,\n",
       "   9,\n",
       "   376,\n",
       "   10,\n",
       "   377,\n",
       "   0.0,\n",
       "   542.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_2|none',\n",
       "   'gi|6666666.146813.572|6666666.146813|NC_014550.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Arthrobacter_arilaitensis_RE117_NC_014550.1',\n",
       "   59.249,\n",
       "   373,\n",
       "   151,\n",
       "   1,\n",
       "   5,\n",
       "   376,\n",
       "   1,\n",
       "   373,\n",
       "   3.7300000000000006e-167,\n",
       "   472.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_2|none',\n",
       "   'gi|6666666.146852.774|6666666.146852|NC_002935.2|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Corynebacterium_diphtheriae_NCTC_13129_NC_002935.2',\n",
       "   61.37,\n",
       "   365,\n",
       "   141,\n",
       "   0,\n",
       "   11,\n",
       "   375,\n",
       "   7,\n",
       "   371,\n",
       "   4.77e-165,\n",
       "   467.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_2|none',\n",
       "   'gi|6666666.146872.438|6666666.146872|NC_017803.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Actinoplanes_sp._SE50/110_NC_017803.1',\n",
       "   57.105,\n",
       "   373,\n",
       "   159,\n",
       "   1,\n",
       "   5,\n",
       "   376,\n",
       "   2,\n",
       "   374,\n",
       "   3.27e-152,\n",
       "   434.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_2|none',\n",
       "   'gi|749927.13.8695|749927.13|NC_014318.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Amycolatopsis_mediterranei_U32_NC_014318.1',\n",
       "   66.938,\n",
       "   369,\n",
       "   122,\n",
       "   0,\n",
       "   8,\n",
       "   376,\n",
       "   5,\n",
       "   373,\n",
       "   0.0,\n",
       "   527.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_2|none',\n",
       "   'gi|930171.11.818|930171.11|NC_015145.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Arthrobacter_phenanthrenivorans_Sphe3_NC_015145.1',\n",
       "   61.518,\n",
       "   369,\n",
       "   141,\n",
       "   1,\n",
       "   9,\n",
       "   376,\n",
       "   6,\n",
       "   374,\n",
       "   1.53e-168,\n",
       "   476.0]],\n",
       " '3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_3|none': [['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_3|none',\n",
       "   'gi|1156913.7.7342|1156913.7|NC_021252.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Amycolatopsis_orientalis_HCCB10007_NC_021252.1',\n",
       "   67.769,\n",
       "   363,\n",
       "   116,\n",
       "   1,\n",
       "   1,\n",
       "   363,\n",
       "   13,\n",
       "   374,\n",
       "   3.0300000000000004e-179,\n",
       "   503.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_3|none',\n",
       "   'gi|169133.3.2262|169133.3|LAB|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Citricoccus_sp_CH26_LAB',\n",
       "   68.32,\n",
       "   363,\n",
       "   115,\n",
       "   0,\n",
       "   1,\n",
       "   363,\n",
       "   12,\n",
       "   374,\n",
       "   0.0,\n",
       "   524.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_3|none',\n",
       "   'gi|290340.24.1033|290340.24|NC_008711.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Arthrobacter_aurescens_TC1_NC_008711.1',\n",
       "   67.493,\n",
       "   363,\n",
       "   118,\n",
       "   0,\n",
       "   1,\n",
       "   363,\n",
       "   12,\n",
       "   374,\n",
       "   0.0,\n",
       "   519.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_3|none',\n",
       "   'gi|367928.21.407|367928.21|NC_008618.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Bifidobacterium_adolescentis_ATCC_15703_NC_008618.1',\n",
       "   57.297,\n",
       "   370,\n",
       "   151,\n",
       "   2,\n",
       "   1,\n",
       "   363,\n",
       "   11,\n",
       "   380,\n",
       "   8.969999999999999e-152,\n",
       "   433.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_3|none',\n",
       "   'gi|442563.11.458|442563.11|AD011|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Bifidobacterium_animalis_subsp._lactis_AD011',\n",
       "   53.243,\n",
       "   370,\n",
       "   166,\n",
       "   2,\n",
       "   1,\n",
       "   363,\n",
       "   10,\n",
       "   379,\n",
       "   1.53e-142,\n",
       "   409.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_3|none',\n",
       "   'gi|446462.15.520|446462.15|NC_021252.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Actinosynnema_mirum_DSM_43827_NC_013093.1_NC_021252.1',\n",
       "   66.942,\n",
       "   363,\n",
       "   119,\n",
       "   1,\n",
       "   1,\n",
       "   363,\n",
       "   16,\n",
       "   377,\n",
       "   0.0,\n",
       "   514.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_3|none',\n",
       "   'gi|452863.24.923|452863.24|NC_011886.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Arthrobacter_chlorophenolicus_A6_NC_011886.1',\n",
       "   68.32,\n",
       "   363,\n",
       "   115,\n",
       "   0,\n",
       "   1,\n",
       "   363,\n",
       "   12,\n",
       "   374,\n",
       "   0.0,\n",
       "   524.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_3|none',\n",
       "   'gi|649831.6.513|649831.6|NC_021191.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Actinoplanes_sp._N902-109_NC_021191.1',\n",
       "   62.81,\n",
       "   363,\n",
       "   135,\n",
       "   0,\n",
       "   1,\n",
       "   363,\n",
       "   11,\n",
       "   373,\n",
       "   4.499999999999999e-170,\n",
       "   479.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_3|none',\n",
       "   'gi|6666666.104540.2193|6666666.104540|AOHO01|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Amycolatopsis_decaplanina_DSM_44594_AOHO01',\n",
       "   67.769,\n",
       "   363,\n",
       "   116,\n",
       "   1,\n",
       "   1,\n",
       "   363,\n",
       "   13,\n",
       "   374,\n",
       "   0.0,\n",
       "   521.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_3|none',\n",
       "   'gi|6666666.111158.8940|6666666.111158|NC_0131311|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Catenulispora_acidiphila_DSM_44928_NC_0131311',\n",
       "   73.003,\n",
       "   363,\n",
       "   98,\n",
       "   0,\n",
       "   1,\n",
       "   363,\n",
       "   10,\n",
       "   372,\n",
       "   0.0,\n",
       "   564.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_3|none',\n",
       "   'gi|6666666.111572.2478|6666666.111572|NC_0094801|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Clavibacter_michiganensis_subsp._michiganensis_NCPPB_382_NC_0094801',\n",
       "   64.641,\n",
       "   362,\n",
       "   126,\n",
       "   2,\n",
       "   1,\n",
       "   362,\n",
       "   302,\n",
       "   661,\n",
       "   1.79e-158,\n",
       "   460.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_3|none',\n",
       "   'gi|6666666.112811.7991|6666666.112811|JOEF01|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Allokutzneria_albata_JOEF01',\n",
       "   67.218,\n",
       "   363,\n",
       "   118,\n",
       "   1,\n",
       "   1,\n",
       "   363,\n",
       "   16,\n",
       "   377,\n",
       "   0.0,\n",
       "   513.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_3|none',\n",
       "   'gi|6666666.146813.572|6666666.146813|NC_014550.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Arthrobacter_arilaitensis_RE117_NC_014550.1',\n",
       "   61.708,\n",
       "   363,\n",
       "   139,\n",
       "   0,\n",
       "   1,\n",
       "   363,\n",
       "   11,\n",
       "   373,\n",
       "   6.15e-169,\n",
       "   476.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_3|none',\n",
       "   'gi|6666666.146852.774|6666666.146852|NC_002935.2|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Corynebacterium_diphtheriae_NCTC_13129_NC_002935.2',\n",
       "   61.326,\n",
       "   362,\n",
       "   139,\n",
       "   1,\n",
       "   1,\n",
       "   362,\n",
       "   11,\n",
       "   371,\n",
       "   7.62e-167,\n",
       "   471.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_3|none',\n",
       "   'gi|6666666.146872.438|6666666.146872|NC_017803.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Actinoplanes_sp._SE50/110_NC_017803.1',\n",
       "   60.055,\n",
       "   363,\n",
       "   145,\n",
       "   0,\n",
       "   1,\n",
       "   363,\n",
       "   12,\n",
       "   374,\n",
       "   1.28e-166,\n",
       "   471.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_3|none',\n",
       "   'gi|749927.13.8695|749927.13|NC_014318.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Amycolatopsis_mediterranei_U32_NC_014318.1',\n",
       "   68.871,\n",
       "   363,\n",
       "   112,\n",
       "   1,\n",
       "   1,\n",
       "   363,\n",
       "   12,\n",
       "   373,\n",
       "   0.0,\n",
       "   527.0],\n",
       "  ['3PGA_AMINOACIDS|2|Phosphoserine_aminotransferase_3|none',\n",
       "   'gi|930171.11.818|930171.11|NC_015145.1|Phosphoserine_aminotransferase_(EC_2.6.1.52)|Arthrobacter_phenanthrenivorans_Sphe3_NC_015145.1',\n",
       "   67.218,\n",
       "   363,\n",
       "   119,\n",
       "   0,\n",
       "   1,\n",
       "   363,\n",
       "   12,\n",
       "   374,\n",
       "   0.0,\n",
       "   518.0]]}"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "best_hit_central_to_orgs_2(blast_File)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "4648fb7d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "119"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test2 = best_hit_central_to_orgs_2(blast_File)\n",
    "tot1 = 0\n",
    "for enz in test2:\n",
    "    tot1 += len(test2[enz])\n",
    "tot1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "89d88bf4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogenase_1|Cglu\\tgi|6666666.146852.1090|6666666.146852|NC_002935.2|D-3-phosphoglycerate_dehydrogenase_(EC_1.1.1.95)|Corynebacterium_diphtheriae_NCTC_13129_NC_002935.2\\t79.623\\t530\\t108\\t0\\t1\\t530\\t1\\t530\\t0.0\\t847.0'"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "li = list(df1.iloc[0])\n",
    "li = [str(el) for el in li]\n",
    "\"\\t\".join(li)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "481355cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "def best_hit_central_to_orgs_3(blast_file):\n",
    "    \"\"\" This function will obtain the best hit from central to organisms. Its output \n",
    "    is a file with these hits. From the output we will obtain the expanded families.\n",
    "    \"\"\"\n",
    "    \n",
    "    # Charge the dataframe with the blastp file:\n",
    "    df = pd.read_csv(blast_file, sep = '\\t', names = CTS.BLAST_COLS)\n",
    "    \n",
    "    ### SE NECESITA QUE LOS ARCHIVOS AUXILIARES ESTÉN CREADOS ###\n",
    "    #Files with the enzimes and orgs info:\n",
    "    \n",
    "    # Make a list with the enzimes:\n",
    "    central_enzimes_file = CTS.BBH_aux_files + 'central_enzimes.txt'\n",
    "    central_enzimes = []\n",
    "    with open(central_enzimes_file, \"r\") as central_file:\n",
    "        for line in central_file:\n",
    "            if line:\n",
    "                central_enzimes.append(line.rstrip(\"\\n\"))\n",
    "    \n",
    "    \n",
    "    # Make a dictionary with the ids and names of organisms:\n",
    "    ids_to_orgs_names_file = CTS.BBH_aux_files + 'ids_to_orgs_names.txt'\n",
    "    genomes = dict()\n",
    "    with open(ids_to_orgs_names_file, \"r\") as orgs_file:\n",
    "        for line in orgs_file:\n",
    "            if line:\n",
    "                key = line.split(\"-->\")[0]\n",
    "                value = line.split(\"-->\")[1]\n",
    "                genomes[key] = value\n",
    "    \n",
    "    \n",
    "    \n",
    "    #Define and initialize the output dictionary:\n",
    "    #unique_central_to_genomes = dict()\n",
    "    #for enzime in central_enzimes:\n",
    "    #    unique_central_to_genomes[enzime] = []\n",
    "     \n",
    "    # Path to best hits file:\n",
    "    \n",
    "    central_to_orgs_bh = CTS.BBH_aux_files + \"central_to_orgs_BH.uniq\"\n",
    "    \n",
    "    \n",
    "    # Finding the best hit:\n",
    "    best_hits_file = open(central_to_orgs_bh, \"w\")\n",
    "    for enzime in central_enzimes:\n",
    "        for genome_id in genomes.keys():\n",
    "            aux_list = []\n",
    "            \n",
    "            for index in range(len(df)):\n",
    "                if (aux_list == []):\n",
    "                    if ((enzime == df[\"query\"][index])&(genome_id in df[\"subject\"][index])):\n",
    "                        aux_list = list(df.iloc[index])\n",
    "                    else:\n",
    "                        continue\n",
    "                else:\n",
    "                    if ((enzime == df[\"query\"][index]) & (genome_id in df[\"subject\"][index]) & (df[\"bitscore\"][index] > aux_list[11])):\n",
    "                        aux_list = list(df.iloc[index])\n",
    "            if aux_list != []: \n",
    "                line = [\"{}\\t\".format(item) for item in aux_list]\n",
    "                best_hits_file.writelines(line)\n",
    "                best_hits_file.write(\"\\n\")\n",
    "                \n",
    "    best_hits_file.close() \n",
    "    #return unique_central_to_genomes\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "31029ae0",
   "metadata": {},
   "outputs": [],
   "source": [
    "best_hit_central_to_orgs_3(blast_File)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b204f3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "cadena = \"Hola\\ta\\ntodos.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ee56ae6",
   "metadata": {},
   "outputs": [],
   "source": [
    "cadena"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "253de13e",
   "metadata": {},
   "source": [
    "### Prueba de escribir best hits segundo blast en un archivo:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "104e667d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def best_hit_orgs_to_central(blast_file):\n",
    "    \"\"\" This function will obtain the best hit from organisms to central, i.e., \n",
    "    assigns to each position id of a blast hit the corresponding enzime that best \n",
    "    hits in that particular place.\n",
    "    \"\"\"\n",
    "    \n",
    "    #charge dataframe:\n",
    "    df = pd.read_csv(blast_file, sep = '\\t', names = CTS.BLAST_COLS, index_col = False)\n",
    "    \n",
    "    #Obtain dictionary with the copy id position to the line of blast\n",
    "    \n",
    "    uniq2 = dict()\n",
    "    for index in df.index:\n",
    "        actual_line = df.iloc[index]\n",
    "        copy_id = actual_line[\"query\"].split(\"|\")[1]  #id of the hit position\n",
    "        if copy_id in uniq2:\n",
    "            if actual_line[\"bitscore\"] > uniq2[copy_id][11]:\n",
    "                uniq2[copy_id] = list(actual_line)\n",
    "        else:\n",
    "            if actual_line[\"bitscore\"] > 0:\n",
    "                uniq2[copy_id] = list(actual_line)\n",
    "    \n",
    "    \n",
    "    # Write the file with the best hits:\n",
    "    \n",
    "    orgs_to_central_bh = CTS.BBH_aux_files + \"orgs_to_central_BH.uniq\"  #Agregar ruta a constantes\n",
    "    best_hits_file = open(orgs_to_central_bh, \"w\")\n",
    "     \n",
    "    for key in uniq2:\n",
    "        line = [\"{}\\t\".format(item) for item in uniq2[key]] \n",
    "        best_hits_file.writelines(line)\n",
    "        best_hits_file.write(\"\\n\")\n",
    "    \n",
    "    \n",
    "    best_hits_file.close()\n",
    "    return\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "c26a16ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_second_blast = '/home/luisyovanny/.evo/dev_Evo/data/data_bases/blastp/genomes_to_central/genomes_to_central.blast'\n",
    "hlp.best_hit_orgs_to_central(path_second_blast)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "e676e6b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "orgs_to_central_bh = CTS.BBH_aux_files + \"orgs_to_central_BH.uniq\"\n",
    "DF1 = pd.read_csv(orgs_to_central_bh, sep = \"\\t\", names = CTS.BLAST_COLS, index_col = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "5ec5c7bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "central_to_orgs_bh = CTS.BBH_aux_files + \"central_to_orgs_BH.uniq\"\n",
    "DF2 = pd.read_csv(central_to_orgs_bh, sep = \"\\t\", names = CTS.BLAST_COLS, index_col = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "87272420",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>subject</th>\n",
       "      <th>pc_identity</th>\n",
       "      <th>aln_length</th>\n",
       "      <th>mismatches</th>\n",
       "      <th>gaps_opened</th>\n",
       "      <th>query_start</th>\n",
       "      <th>query_end</th>\n",
       "      <th>subject_start</th>\n",
       "      <th>subject_end</th>\n",
       "      <th>e_value</th>\n",
       "      <th>bitscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogena...</td>\n",
       "      <td>gi|1156913.7.6305|1156913.7|NC_021252.1|D-3-ph...</td>\n",
       "      <td>59.203</td>\n",
       "      <td>527</td>\n",
       "      <td>214</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>526</td>\n",
       "      <td>1</td>\n",
       "      <td>527</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>613.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogena...</td>\n",
       "      <td>gi|169133.3.2692|169133.3|LAB|D-3-phosphoglyce...</td>\n",
       "      <td>46.183</td>\n",
       "      <td>524</td>\n",
       "      <td>280</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>523</td>\n",
       "      <td>1</td>\n",
       "      <td>523</td>\n",
       "      <td>2.160000e-155</td>\n",
       "      <td>455.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogena...</td>\n",
       "      <td>gi|290340.24.2545|290340.24|NC_008711.1|D-3-ph...</td>\n",
       "      <td>49.049</td>\n",
       "      <td>526</td>\n",
       "      <td>267</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>530</td>\n",
       "      <td>5</td>\n",
       "      <td>530</td>\n",
       "      <td>2.060000e-161</td>\n",
       "      <td>470.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogena...</td>\n",
       "      <td>gi|367928.21.1188|367928.21|NC_008618.1|D-3-ph...</td>\n",
       "      <td>34.098</td>\n",
       "      <td>305</td>\n",
       "      <td>182</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>299</td>\n",
       "      <td>6</td>\n",
       "      <td>303</td>\n",
       "      <td>1.040000e-41</td>\n",
       "      <td>155.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogena...</td>\n",
       "      <td>gi|442563.11.829|442563.11|AD011|D-3-phosphogl...</td>\n",
       "      <td>36.749</td>\n",
       "      <td>283</td>\n",
       "      <td>171</td>\n",
       "      <td>3</td>\n",
       "      <td>15</td>\n",
       "      <td>292</td>\n",
       "      <td>13</td>\n",
       "      <td>292</td>\n",
       "      <td>3.030000e-46</td>\n",
       "      <td>167.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               query  \\\n",
       "0  3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogena...   \n",
       "1  3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogena...   \n",
       "2  3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogena...   \n",
       "3  3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogena...   \n",
       "4  3PGA_AMINOACIDS|1|Phosphoglycerate_dehydrogena...   \n",
       "\n",
       "                                             subject  pc_identity  aln_length  \\\n",
       "0  gi|1156913.7.6305|1156913.7|NC_021252.1|D-3-ph...       59.203         527   \n",
       "1  gi|169133.3.2692|169133.3|LAB|D-3-phosphoglyce...       46.183         524   \n",
       "2  gi|290340.24.2545|290340.24|NC_008711.1|D-3-ph...       49.049         526   \n",
       "3  gi|367928.21.1188|367928.21|NC_008618.1|D-3-ph...       34.098         305   \n",
       "4  gi|442563.11.829|442563.11|AD011|D-3-phosphogl...       36.749         283   \n",
       "\n",
       "   mismatches  gaps_opened  query_start  query_end  subject_start  \\\n",
       "0         214            1            1        526              1   \n",
       "1         280            2            1        523              1   \n",
       "2         267            1            6        530              5   \n",
       "3         182            6            7        299              6   \n",
       "4         171            3           15        292             13   \n",
       "\n",
       "   subject_end        e_value  bitscore  \n",
       "0          527   0.000000e+00     613.0  \n",
       "1          523  2.160000e-155     455.0  \n",
       "2          530  2.060000e-161     470.0  \n",
       "3          303   1.040000e-41     155.0  \n",
       "4          292   3.030000e-46     167.0  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "0db0201a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def bbh_detector(central_to_orgs_bhs, orgs_to_central_bhs):\n",
    "    \"\"\" This function obtains the best bidirectional hits between\n",
    "    the blast of central to organisms and organisms to central respectively.\n",
    "    \"\"\"\n",
    "\n",
    "    # charge dataframes:\n",
    "    \n",
    "    df = pd.read_csv(central_to_orgs_bhs, sep = ' ', names = CTS.BLAST_COLS, index_col = False)\n",
    "    \n",
    "    df1 = pd.read_csv(orgs_to_central_bhs, sep = ' ', names = CTS.BLAST_COLS, index_col = False)\n",
    "    \n",
    "    # Write the file with the bbh:\n",
    "    \n",
    "    BBH_file = CTS.BBH_aux_files + \"BBH.bbh\"  #Agregar ruta a constantes\n",
    "    bbh_file = open(BBH_file, \"w\")\n",
    "     \n",
    "    for indx in df1.index:\n",
    "        copy_id = df1.iloc[indx][\"query\"].split(\"|\")[1]\n",
    "        enzime = df1.iloc[indx][\"subject\"]\n",
    "        for indxx in df.index:\n",
    "            if copy_id not in df.iloc[indxx][\"subject\"]:\n",
    "                continue\n",
    "            elif enzime not in df.iloc[indxx][\"query\"]:\n",
    "                continue\n",
    "            else:\n",
    "                #bbh[copy_id] = enzime\n",
    "                bbh_file.write(copy_id + \" -> \")\n",
    "                bbh_file.write(enzime +\"\\t\")\n",
    "                bbh_file.write(df1.iloc[indx][\"query\"]+\"\\n\")\n",
    "           \n",
    "    \n",
    "    bbh_file.close()\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "a7a20107",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/luisyovanny/.evo/dev_Evo/evo_helpers.py:576: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  enzime = df1.iloc[indx][1]\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "argument of type 'numpy.float64' is not iterable",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[54], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m cent_to_org_bh \u001b[38;5;241m=\u001b[39m CTS\u001b[38;5;241m.\u001b[39mBBH_aux_files \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcentral_to_orgs_BH.uniq\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m      2\u001b[0m org_to_cent_bh \u001b[38;5;241m=\u001b[39m CTS\u001b[38;5;241m.\u001b[39mBBH_aux_files \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124morgs_to_central_BH.uniq\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m----> 3\u001b[0m \u001b[43mhlp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbbh_detector\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcent_to_org_bh\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43morg_to_cent_bh\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      4\u001b[0m \u001b[38;5;66;03m#df1 = pd.read_csv(org_to_cent_bh, sep = ' ', names = CTS.BLAST_COLS, index_col =False)\u001b[39;00m\n",
      "File \u001b[0;32m~/.evo/dev_Evo/evo_helpers.py:578\u001b[0m, in \u001b[0;36mbbh_detector\u001b[0;34m(central_to_orgs_bhs, orgs_to_central_bhs)\u001b[0m\n\u001b[1;32m    576\u001b[0m enzime \u001b[38;5;241m=\u001b[39m df1\u001b[38;5;241m.\u001b[39miloc[indx][\u001b[38;5;241m1\u001b[39m]\n\u001b[1;32m    577\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m indxx \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mlist\u001b[39m(df\u001b[38;5;241m.\u001b[39mindex):\n\u001b[0;32m--> 578\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[43mcopy_id\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mnot\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mdf\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43miloc\u001b[49m\u001b[43m[\u001b[49m\u001b[43mindxx\u001b[49m\u001b[43m]\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43msubject\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m:\n\u001b[1;32m    579\u001b[0m         \u001b[38;5;28;01mcontinue\u001b[39;00m\n\u001b[1;32m    580\u001b[0m     \u001b[38;5;28;01melif\u001b[39;00m enzime \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m df\u001b[38;5;241m.\u001b[39miloc[indxx][\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mquery\u001b[39m\u001b[38;5;124m\"\u001b[39m]:\n",
      "\u001b[0;31mTypeError\u001b[0m: argument of type 'numpy.float64' is not iterable"
     ]
    }
   ],
   "source": [
    "cent_to_org_bh = CTS.BBH_aux_files + \"central_to_orgs_BH.uniq\"\n",
    "org_to_cent_bh = CTS.BBH_aux_files + \"orgs_to_central_BH.uniq\"\n",
    "hlp.bbh_detector(cent_to_org_bh, org_to_cent_bh)\n",
    "#df1 = pd.read_csv(org_to_cent_bh, sep = ' ', names = CTS.BLAST_COLS, index_col =False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d44656ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27ec7b45",
   "metadata": {},
   "outputs": [],
   "source": [
    "import evo_helpers as hlp\n",
    "bbh_detector(cent_to_org_bh, org_to_cent_bh)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5f67561",
   "metadata": {},
   "source": [
    "## Comparar los bbh contra los obtenidos en evomining:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69f47e0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "BBH_mio = list()\n",
    "BBH_mio_path = \"/home/luisyovanny/.evo/dev_Evo/data/aux_data/bbh_aux_files/BBH.bbh\"\n",
    "bbh_mio = open(BBH_mio_path, 'r')\n",
    "for line in bbh_mio:\n",
    "    hit = line.split(\"\\t\")[0]\n",
    "    BBH_mio.append(hit)\n",
    "\n",
    "\n",
    "bbh_mio.close()\n",
    "BBH_mio_set = set(BBH_mio)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb35e289",
   "metadata": {},
   "outputs": [],
   "source": [
    "BBH_evo = list()\n",
    "BBH_evo_path = \"/home/luisyovanny/.evo/dev_Evo/vueltalos17.blast.bbh\"\n",
    "bbh_evo = open(BBH_evo_path, 'r')\n",
    "for line in bbh_evo:\n",
    "    hit = line.split(\"\\t\")[0]\n",
    "    BBH_evo.append(hit)\n",
    "\n",
    "\n",
    "bbh_evo.close()\n",
    "BBH_evo_set = set(BBH_evo)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c83633e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "BBH_evo_set == BBH_mio_set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cef4efd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#bbh_set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b088d2f1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9de33f94",
   "metadata": {},
   "outputs": [],
   "source": [
    "expanded_families_files = [x for x in os.listdir(CTS.EXPANDED_FAMS) if x.endswith(\".fasta\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8d13002",
   "metadata": {},
   "outputs": [],
   "source": [
    "expanded_families_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f10b66b",
   "metadata": {},
   "outputs": [],
   "source": [
    "blastdb_path = CTS.BLASTDBs_PATH + \"nat_prods_blastdb/\"\n",
    "h = blastdb_path + blastdb_path.split(\"/\")[-2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55d5ebe3",
   "metadata": {},
   "outputs": [],
   "source": [
    "h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d25c2ab2",
   "metadata": {},
   "outputs": [],
   "source": [
    "blastdb_path = CTS.BLASTDBs_PATH + 'central_blastdb/'\n",
    "h = blastdb_path + blastdb_path.split(\"/\")[-2]\n",
    "h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8518051",
   "metadata": {},
   "outputs": [],
   "source": [
    "mibig_db = CTS.MIBiG_NAT_PRODS_DB + \"nat_prods.fasta\"\n",
    "mibig_file = open(mibig_db, \"r\")\n",
    "line = mibig_file.readline()\n",
    "li = line.strip()\n",
    "print(li)\n",
    "mibig_file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c8dfe1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "k = \"hola\"\n",
    "\"h\" in k"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7dc4ce5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def std_highlighter(x):\n",
    "    style_lt = \"background-color: white; color: black; font-weight: bold; border: solid; text-align: center; font-size:1.2em;\"\n",
    "    style_gt = \"background-color: #C04000; color: black; font-weight: bold; border: solid; text-align:center;\"\n",
    "    gt_std = x > x.mean() + x.std()\n",
    "    return [style_gt if i else style_lt for i in gt_std]\n",
    "\n",
    "def std2_highlighter(x):\n",
    "    style_lt = \"background-color: white; color: black; font-weight: bold; border: solid; text-align: center; font-size:1.2em;\"\n",
    "    style_gt = \"background-color: #C04000; color: black; font-weight: bold; border: solid; text-align:center;\"\n",
    "    gt_std = x > x.mean() + 2*x.std()\n",
    "    return [style_gt if i else style_lt for i in gt_std]\n",
    "\n",
    "def mean_highlighter(x):\n",
    "    style_lt = \"background-color: white; color: black; font-weight: bold; border: solid; text-align: center; font-size:1.2em;\"\n",
    "    style_gt = \"background-color: #C04000; color: black; font-weight: bold; border: solid; text-align:center;\"\n",
    "    gt_mean = x > x.mean()\n",
    "    return [style_gt if i else style_lt for i in gt_mean]\n",
    "def median_highlighter(x):\n",
    "    style_lt = \"background-color: white; color: black; font-weight: bold; border: solid; text-align: center; font-size:1.2em;\"\n",
    "    style_gt = \"background-color: #C04000; color: black; font-weight: bold; border: solid; text-align:center;\"\n",
    "    gt_median = x > x.mean()\n",
    "    return [style_gt if i else style_lt for i in gt_median]\n",
    "\n",
    "def mode_highlighter(x):\n",
    "    import statistics as st\n",
    "    style_lt = \"background-color: white; color: black; font-weight: bold; border: solid; text-align: center; font-size:1.2em;\"\n",
    "    style_gt = \"background-color: #C04000; color: black; font-weight: bold; border: solid; text-align:center;\"\n",
    "    gt_mode = x > st.mode(x)\n",
    "    return [style_gt if i else style_lt for i in gt_mode]\n",
    "\n",
    "def genomes_highlighter(x):\n",
    "    style_genome = \"background-color: white; color: black; font-weight: bold; border: solid; text-align: center; font-size:1.2em;\"\n",
    "    return [style_genome for i in x]\n",
    "headers = {\n",
    "    'selector': 'th.col_heading',\n",
    "    'props': \"background-color: OldLace; color: black; font-weight: bold; border: solid; text-align: center; font-size:1.2em;\"\n",
    "} \n",
    "def apply_highlighter2(df_heat,function, style_highlighter, headers_style):\n",
    "    return df_heat.style.apply(function,subset=[\"fam_1\", \"fam_2\"]).apply(style_highlighter, subset=\"orgs_ids\").hide(axis=\"index\").set_table_styles([headers_style], overwrite = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "349c8521",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_heat_map(criteria):\n",
    "    import json\n",
    "    \n",
    "    json_file_Expfams = open(CTS.EXP_FAMS_JSON, \"r\")\n",
    "    exp_fam_dictio = json.load(json_file_Expfams)\n",
    "    json_file_Expfams.close()\n",
    "    \n",
    "    json_file_org = open(CTS.ORGS_IDS_NMS, \"r\")\n",
    "    orgs_ids_dict = json.load(json_file_org)\n",
    "    orgs_ids_list = list(orgs_ids_dict.keys())\n",
    "    json_file_org.close()\n",
    "    \n",
    "    json_file_fams = open(CTS.FAMILIES, \"r\")\n",
    "    fams_dict = json.load(json_file_fams)\n",
    "    families = list(fams_dict.keys())\n",
    "    json_file_fams.close()\n",
    "    \n",
    "    Dictio_DF = {\"orgs_ids\": orgs_ids_list}\n",
    "    for fam in families:\n",
    "        Dictio_DF[fam] = []\n",
    "    for org in Dictio_DF[\"orgs_ids\"]:\n",
    "        for key in exp_fam_dictio:\n",
    "            if org in key:\n",
    "                fmly = key.split(\"|\")[0]\n",
    "                Dictio_DF[fmly].append(len(exp_fam_dictio[key]))\n",
    "    \n",
    "    #Dictio_DF\n",
    "    DF_heat = pd.DataFrame.from_dict(Dictio_DF)\n",
    "    #for indx in DF_heat.index:\n",
    "    #    Id = DF_heat[\"organisms\"][indx]\n",
    "    #    DF_heat.at[indx,\"organisms\"] = orgs_ids_dict[Id]\n",
    "    \n",
    "    \n",
    "    #DF_heat\n",
    "    headers_style = {\n",
    "    'selector': 'th.col_heading',\n",
    "    'props': \"background-color: OldLace; color: black; font-weight: bold; border: solid; text-align: center; font-size:1.2em;\"\n",
    "    }\n",
    "    if criteria == \"mean\":\n",
    "        return apply_highlighter2(DF_heat,mean_highlighter, genomes_highlighter, headers_style)\n",
    "    elif criteria == \"median\":\n",
    "        return apply_highlighter2(DF_heat,median_highlighter, genomes_highlighter, headers_style)\n",
    "    elif criteria == \"mean + std\":\n",
    "        return apply_highlighter2(DF_heat,std_highlighter, genomes_highlighter, headers_style)\n",
    "    elif criteria == \"mean + 2*std\":\n",
    "        return apply_highlighter2(DF_heat,std2_highlighter, genomes_highlighter, headers_style)\n",
    "    elif criteria == \"mode\":\n",
    "        return apply_highlighter2(DF_heat,mode_highlighter, genomes_highlighter, headers_style)\n",
    "    else: \n",
    "        return print(f\"{criteria} is not a valid criteria\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eede47e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "make_heat_map(\"mean + std\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c852b0b",
   "metadata": {},
   "source": [
    "## MUSCLE:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4c861739",
   "metadata": {},
   "outputs": [],
   "source": [
    "import evo_helpers as hlp\n",
    "import evo_constants as CTS\n",
    "from Bio.Align.Applications import MuscleCommandline\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d7234e99",
   "metadata": {},
   "outputs": [],
   "source": [
    "muscle_exe = \"/home/luisyovanny/muscle5.1.linux_intel64\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c72c1a43",
   "metadata": {},
   "outputs": [],
   "source": [
    "in_file = CTS.EXPANDED_FAMS + \"expanded_fam_1.fasta\"\n",
    "out_file = CTS.MUSCLE_OUTPUT + \"MUSCLE_\" + \"expanded_fam_1.fasta\"\n",
    "#muscle_cline = MuscleCommandline(muscle_exe, input=in_file, out=out_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "5156f545",
   "metadata": {},
   "outputs": [],
   "source": [
    "#muscle_cline = MuscleCommandline(muscle_exe, input = in_file, out = out_file)\n",
    "import subprocess\n",
    "#subprocess.call('/home/csar/muscle_v5/muscle5.1.linux_intel64 -align %s -output %s'%(in_file,out_file))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ce90a1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#import os\n",
    "#os.popen('/home/csar/muscle_v5/muscle5.1.linux_intel64 -align %s -output %s'%(in_file,out_file))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "799a63be",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "subprocess.run([\"/home/luisyovanny/muscle5.1.linux_intel64\", \"-align\", in_file, \"-output\", out_file])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed434084",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
